diff -crbBN BitchX/Changes.esad BitchX.esad/Changes.esad
*** BitchX/Changes.esad	Wed Dec 31 19:00:00 1969
--- BitchX.esad/Changes.esad	Mon Jan 13 19:24:43 1997
***************
*** 0 ****
--- 1,34 ----
+ esad-4  (1/  /1997)
+ - Added version masking support
+ 	- when someone does a version on you the client will randomly choose
+ 	  a version from ~/.BitchX/BitchX.versions
+ 	  see the file BitchX.versions in the CubicZ/ dir for an example
+ 	- NOTE: this did require some rewriting of do_version()
+ - Changed the return types of main() in ircflush.c and wserv.c
+ 	- I was sick of seeing the compiler warnings
+ - Fixed problems with certain numerics that could be abused by servers to
+   cause segfaults.
+ - Added /ehelp it's the same as /bhelp but it's esad specific stuff.
+   Thanks to panasync for making chelp so portable.
+ - Changed equal_nickname in dcc.c to 34 instead of 20
+   (don't know what it does but panasync pointed it out)
+ 
+ esad-3	(1/6/1997)
+ - Added a bunch of *serv functions...maybe i'll write them out later
+ 
+ esad-2	(1/3/1997)
+ - first PUBLIC release 
+ - official dist site ftp://ftp.underworld.net/pub/irc/clients/BitchX/
+ - Added FORMAT_SERVER_UNOTICE
+ 	-FORMAT_SERVER_UNOTICE standard unotice
+ 	-FORMAT_SERVER_UNOTICE_CRASH for when someone sends a crash to
+ 		someone
+ 	-FORMAT_SERVER_UNOTICE_JUPE when someone jupes a server
+ 	-FORMAT_SERVER_UNOTICE_KILL when someone kills someone with
+ 		IceServ
+ 	-FORMAT_SERVER_UNOTICE_REJECT when a host is rejected
+ - fixed bug with not defining DEFAULT_ESAD_FAKE_HOST in include/config.h
+ 
+ esad-1	(?/?/1996)
+ - Initial release
+ - fake host support
diff -crbBN BitchX/CubicZ/BitchX.versions BitchX.esad/CubicZ/BitchX.versions
*** BitchX/CubicZ/BitchX.versions	Wed Dec 31 19:00:00 1969
--- BitchX.esad/CubicZ/BitchX.versions	Sun Jan 12 12:25:26 1997
***************
*** 0 ****
--- 1,27 ----
+ BitchX-68d by panasync - Linux 2.0.27 : Keep it to yourself!
+ .x<acid!irc>x. v6.00 by acidjazz, simply untouchable
+ ComBot Version 2.05
+ [PurePak] v1.08 by Crypt Keeper / -+ TEXT +-
+ Dough's vms client:1.5(3.0.7r bighorn):VMS
+ PLuToNiuM IRC Client - Accept no imitations :ScriptBabe Omega4 By SpErM!fzeidan@utep.edu
+ VERSION - unknown command.
+ WSIRC 1.12-a - Copyright 1994 Caesar M Samsi csamsi@clark.net TEXT CHANNEL
+ ircII 2.2.9 *IX :DETURBO 3.1
+ ircII 2.2.9.WT+ *IX :Are we there yet ?
+ ircII 2.6 SunOS 4.1.3 :ircii 2.6: almost there...
+ ircII 2.8-EPIC2 Linux 2.0.23 -  ircII2.8's only botrunner's client.
+ ircII 2.8.2 FreeBSD 2.1.0-RELEASE :this is a bug free client.  honest
+ ircII 2.8.2 Linux 2.0.25 :ircii 2.8: almost there...
+ ircII 2.8.2-EPIC3pre5 Linux 1.2.1 - All the things phone wont include.....
+ ircII 2.9-BitchX-60 Linux 1.2.8 :bitZ%summer '96(bitX%summer'96)
+ mIRC16 v3.64 IRC Client K.Mardam-Bey
+ mIRC16 v3.8 IRC Client K.Mardam-Bey
+ mIRC16 v3.92 IRC Client K.Mardam-Bey
+ mIRC16 v4.5 IRC Client K.Mardam-Bey
+ mIRC16 v4.6 IRC Client K.Mardam-Bey
+ mIRC32 v4.0 IRC Client K.Mardam-Bey
+ mIRC32 v4.5 IRC Client K.Mardam-Bey
+ mIRC32 v4.52 IRC Client K.Mardam-Bey
+ mIRC32 v4.6 IRC Client K.Mardam-Bey
+ mIRC32 v4.7 K.Mardam-Bey
+ zap32 4.7 by zap (efnet)
diff -crbBN BitchX/CubicZ/esad.help BitchX.esad/CubicZ/esad.help
*** BitchX/CubicZ/esad.help	Wed Dec 31 19:00:00 1969
--- BitchX.esad/CubicZ/esad.help	Mon Jan 13 18:17:28 1997
***************
*** 0 ****
--- 1,57 ----
+ INDEX
+    %YChanServ commands:%n
+       %BCSACCESS  CSADD     CSDEL     CSHELP    CSINFO     CSINVITE CSREGISTER%n
+       %BCSUNBAN   CSUP%n
+    %YNickServ commands:%n
+       %BNSADD     NSGHOST   NSIDENT   NSRECOVER NSREGISTER NSSET%n
+    %YOperServ commands:%n
+       %BCRASH     GLINE     IADD      IDEL      IINVITE    IKICK    IKILL%n
+       %BILIST     IMODE     INICK     INICKLIST IUNBAN     JUPE%n
+ HELP
+    %GUsage%n: %W/%nehelp %Y<%nTopic%G|%nIndex%Y>%n
+      %YTopic%n - This gives help on %Y<%nTopic%Y>%n
+      %YIndex%n - This shows the list of commands supplied in esad
+ CSACCESS
+    %GUsage%n: %W/%ncsaccess %Y[%n#channel%Y]%n nick
+      Shows you %Rnick's%n access on %R#channel%n
+ CSADD
+    %GUsage%n: %W/%ncsadd %Y[%n#channel%Y]%n nick level
+      Adds %Rnick%n to %R#channel%n with %Rlevel%n access
+ CSDEL
+    %GUsage%n: %W/%ncsadd %Y[%n#channel%Y]%n nick
+      Deletes %Rnick%n from %R#channel%n
+ CSHELP
+    %GUsage%n: %W/%ncshelp %Y[%ntopic%Y]%n
+      Gives you help on %Rtopic%n
+ CSINFO
+    %GUsage%n: %W/%ncsinfo %Y[%n#channel%Y]%n
+      Gives you info on %R#channel%n
+ CSINVITE
+    %GUsage%n: %W/%ncsinvite #channel
+      Invites you to %R#channel%n
+ CSREGISTER
+    %GUsage%n: %W/%ncsregister %Y[%n#channel%Y]%n password
+      Registers %R#channel%n with ChanServ using %Rpassword%n
+ CSUNBAN
+    %GUsage%n: %W/%ncsunban #channel
+      Unbans you on %R#channel%n
+ CSUP
+    %GUsage%n: %W/%ncsup %Y[%n#channel%Y]%n
+      Ops you on %R#channel%n
+ NSADD
+    %GDISABLED FOR NOW%n 
+ NSGHOST
+    %GUsage%n: %W/%nnsghost nick
+      Kills ghosted %Rnick%n
+ NSIDENT
+    %GUsage%n: %W/%nnsident password
+      Identifies you using %Rpassword%n
+ NSRECOVER
+    %GUsage%n: %W/%nnsrecover nick
+      Kills %Rnick%n if you are authorized to do so.
+ NSREGISTER
+    %GUsage%n: %W/%nnsregister password
+      Registers your nickname with NickServ using %Rpassword%n
+ NSSET
+    %GUsage%n: %W/%nnsset option value
+      Changes NickServ setting %Roption%n to %Rvalue%n
diff -crbBN BitchX/INSTALLATION BitchX.esad/INSTALLATION
*** BitchX/INSTALLATION	Tue Nov 12 23:03:57 1996
--- BitchX.esad/INSTALLATION	Mon Jan 13 19:38:01 1997
***************
*** 81,86 ****
--- 81,88 ----
        ifying /set CTOOLZ_DIR = some_other_dir in your .ircrc file. Some 
        scripts will alter this setting upon loading.
  
+       NOTE: for esad you should also copy esad.help from CubicZ/ to ~/.BitchX
+ 
  [~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~]
  
  Machine Specific Instructions
diff -crbBN BitchX/QUIKINST BitchX.esad/QUIKINST
*** BitchX/QUIKINST	Tue Nov 12 23:03:57 1996
--- BitchX.esad/QUIKINST	Mon Jan 13 19:38:01 1997
***************
*** 81,86 ****
--- 81,88 ----
        ifying /set CTOOLZ_DIR = some_other_dir in your .ircrc file. Some 
        scripts will alter this setting upon loading.
  
+       NOTE: for esad you should also copy esad.help from CubicZ/ to ~/.BitchX
+ 
  [~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~][~]
  
  Machine Specific Instructions
diff -crbBN BitchX/README.esad BitchX.esad/README.esad
*** BitchX/README.esad	Wed Dec 31 19:00:00 1969
--- BitchX.esad/README.esad	Sun Jan 12 12:37:01 1997
***************
*** 0 ****
--- 1,11 ----
+ This is a patch for BitchX written by SPaM (spam@underworld.net).
+ It is mainly for use on the EdNet irc network (irc.underworld.net) but it
+ does have a few features that will work regardless of which net you are
+ on.  So far most of the changes are either for IRC ops on EdNet or for the
+ use of services on EdNet.  If you have any questions/bugs email
+ spam@underworld.net or contact SPaM on #couzin-ed at irc.underworld.net.
+ Please do NOT bug panasync with any bugs since he won't be able todo
+ anything.  One final note:  I have no problem with you using my code, just
+ please give me credit for it.
+ 
+ SPaM
diff -crbBN BitchX/include/config.h BitchX.esad/include/config.h
*** BitchX/include/config.h	Wed Dec 25 23:24:15 1996
--- BitchX.esad/include/config.h	Mon Jan 13 19:33:20 1997
***************
*** 25,31 ****
   */
  
  #ifndef DEFAULT_SERVER
! #define DEFAULT_SERVER	    "irc.limited.net"
  #endif
  
  /*
--- 25,31 ----
   */
  
  #ifndef DEFAULT_SERVER
! #define DEFAULT_SERVER	    "irc.underworld.net"
  #endif
  
  /*
***************
*** 356,363 ****
  #define DEFAULT_STATUS_OPER_KILLS "[nk %K|ok %K]"
  #define DEFAULT_STATUS_USERS "[O:%! N:%! I:%! V:%! F:%!]"
  
! #define DEFAULT_USERMODE "+iw"  /* change this to the default usermode */
! #define DEFAULT_OPERMODE "swfck"
  
  #undef ISP_VERSION   /* define this if something offensive offends you
  		      * like BitchX Roqz yur NutZ an stuff
--- 356,363 ----
  #define DEFAULT_STATUS_OPER_KILLS "[nk %K|ok %K]"
  #define DEFAULT_STATUS_USERS "[O:%! N:%! I:%! V:%! F:%!]"
  
! #define DEFAULT_USERMODE "+i"  /* change this to the default usermode */
! #define DEFAULT_OPERMODE "crkfn"
  
  #undef ISP_VERSION   /* define this if something offensive offends you
  		      * like BitchX Roqz yur NutZ an stuff
***************
*** 410,414 ****
--- 410,437 ----
  #else
  #define DEFAULT_DCC_FAST 0
  #endif
+ 
+ /* Begin defines added by SPaM (spam@underworld.net) */
+ 
+ /* Fake host support for the +esad irc server
+  * If you do not need this support simply change the #define to #undef.
+  * NOTE this should NOT be set to NULL or else you may have problems 
+  * connecting to irc.
+  */
+ 
+ #undef	DEFAULT_ESAD_FAKE_HOST "frosh.blackness.net"
+ #define	DEFAULT_ESAD_FAKE_HOST_PASS "PaSSwORd"
+ 
+ /* Default nicknames for the irc services on EdNet (irc.underworld.net) */
+ 
+ #define DEFAULT_CHANSERV "ChanServ"
+ #define DEFAULT_MEMOSERV "MemoServ"
+ #define DEFAULT_NICKSERV "NickServ"
+ #define DEFAULT_OPERSERV "IceServ"
+ 
+ /* Default path for esad help file */
+ /* again thanks to pana for the great system */
+ 
+ #define DEFAULT_ESAD_HELP_FILE DEFAULT_CTOOLZ_DIR"/esad.help"
  
  #endif /* __config_h_ */
diff -crbBN BitchX/include/esad.h BitchX.esad/include/esad.h
*** BitchX/include/esad.h	Wed Dec 31 19:00:00 1969
--- BitchX.esad/include/esad.h	Sun Jan 12 12:01:30 1997
***************
*** 0 ****
--- 1 ----
+ char *get_version(void);
diff -crbBN BitchX/include/irc.h BitchX.esad/include/irc.h
*** BitchX/include/irc.h	Tue Dec 24 16:18:40 1996
--- BitchX.esad/include/irc.h	Fri Jan  3 15:29:24 1997
***************
*** 165,171 ****
  # define DAEMON_UID 1
  #endif
  
! #define NICKNAME_LEN 9
  #define NAME_LEN 80
  #define REALNAME_LEN 50
  #define PATH_LEN 1024
--- 165,171 ----
  # define DAEMON_UID 1
  #endif
  
! #define NICKNAME_LEN 20
  #define NAME_LEN 80
  #define REALNAME_LEN 50
  #define PATH_LEN 1024
diff -crbBN BitchX/include/output.h BitchX.esad/include/output.h
*** BitchX/include/output.h	Sat Dec 14 23:42:42 1996
--- BitchX.esad/include/output.h	Tue Jan 14 15:36:13 1997
***************
*** 18,23 ****
--- 18,24 ----
  	void	send_to_server _((char *, ...));
  	void	say _((char *, ...));
  	void	bitchsay _((char *, ...));
+ 	void	esadsay _((char *, ...));
  	void	serversay _((int, char *, ...));
  	void	yell _((char *, ...));
  	void	help_put_it _((char *, char *, ...));
***************
*** 25,31 ****
  	void	put_it _(());
  	void	send_to_server _(());
  	void	say _(());
! 	void	bitchsay _(());
  	void	yell _(());
  	void	help_put_it _(());
  #endif /* HAVE_STDARG_H */
--- 26,32 ----
  	void	put_it _(());
  	void	send_to_server _(());
  	void	say _(());
! 	void	esadsay _(());
  	void	yell _(());
  	void	help_put_it _(());
  #endif /* HAVE_STDARG_H */
diff -crbBN BitchX/include/vars.h BitchX.esad/include/vars.h
*** BitchX/include/vars.h	Fri Dec 27 23:23:35 1996
--- BitchX.esad/include/vars.h	Tue Jan 14 15:33:14 1997
***************
*** 79,84 ****
--- 79,85 ----
  	CD_DEVICE_VAR,
  #endif
  	CHANNEL_NAME_WIDTH_VAR ,
+ 	CHANSERV_VAR,
  	CHECK_BEEP_USERS_VAR ,
  	CLIENTINFO_VAR ,
  	CLOAK_VAR ,
***************
*** 113,118 ****
--- 114,124 ----
  	EGO_VAR ,
  	EIGHT_BIT_CHARACTERS_VAR,
  	ENCRYPT_PROGRAM_VAR ,
+ #ifdef DEFAULT_ESAD_FAKE_HOST
+ 	ESAD_FAKE_HOST_VAR ,
+ 	ESAD_FAKE_HOST_PASS_VAR ,
+ #endif
+ 	ESAD_HELP_VAR,
  	EXEC_PROTECTION_VAR ,
  	FAKE_HOST_VAR , 
  	FAKE_SPLIT_PATS_VAR,
***************
*** 281,286 ****
--- 287,297 ----
  FORMAT_SERVER_NOTICE_TRAFFIC_HIGH_VAR,
  FORMAT_SERVER_NOTICE_TRAFFIC_NORM_VAR,
  FORMAT_SERVER_NOTICE_UNAUTH_VAR,
+ FORMAT_SERVER_UNOTICE_VAR,
+ FORMAT_SERVER_UNOTICE_CRASH_VAR,
+ FORMAT_SERVER_UNOTICE_JUPE_VAR,
+ FORMAT_SERVER_UNOTICE_KILL_VAR,
+ FORMAT_SERVER_UNOTICE_REJECT_VAR,
  
  FORMAT_SET_VAR,
  FORMAT_SET_NOVALUE_VAR,
***************
*** 392,397 ****
--- 403,409 ----
  	MAX_DEOPS_VAR ,
  	MAX_IDLEKICKS_VAR ,
  	MAX_RECURSIONS_VAR ,
+ 	MEMOSERV_VAR,
  	MENU_VAR ,
  	MIRCS_VAR,
  	MODE_STRIPPER_VAR ,
***************
*** 403,408 ****
--- 415,421 ----
  	NICK_COMPLETION_TYPE_VAR,
  	NICKFLOOD_VAR,
  	NICKFLOOD_TIME_VAR,
+ 	NICKSERV_VAR,
  	NO_CTCP_FLOOD_VAR,
  	NOTIFY_HANDLER_VAR,
  	NOTIFY_LEVEL_VAR,
***************
*** 411,416 ****
--- 424,430 ----
  	NUM_BANMODES_VAR,
  	NUM_OPMODES_VAR,
  	OPER_MODES_VAR,
+ 	OPERSERV_VAR,
  	OV_VAR,
  	PAD_CHAR_VAR,
  	PING_TYPE_VAR,
***************
*** 586,591 ****
--- 600,606 ----
  extern	int	loading_global;
  	int cget_int_var _((enum CVAR_TYPES, char *));
  	void	savebitchx_variables _((FILE *));
+ 	void	saveesad_variables _((FILE *));
  	void	reinit_autoresponse _((Window *, char *, int));
  	void	clear_sets _((void));
  	void	clear_bindings _((void));
diff -crbBN BitchX/source/Makefile.in BitchX.esad/source/Makefile.in
*** BitchX/source/Makefile.in	Thu Dec 26 00:42:46 1996
--- BitchX.esad/source/Makefile.in	Sun Jan  5 15:27:29 1997
***************
*** 21,27 ****
  	pluto.o queue.o readlog.o reg.o scandir.o screen.o server.o\
  	stack.o status.o term.o timer.o translat.o userlist.o vars.o\
  	whois.o whowas.o window.o words.o bot_link.o struct.o dich_conf.o\
! 	$(PP_OBJS) $(TCL_OBJS) $(CD_OBJS)
  
  SOURCES = alias.c array.c banlist.c bot.c cdcc.c chelp.c crypt.c ctcp.c\
  	dcc.c edit.c edit2.c exec.c flood.c glob.c files.c funny.c help.c\
--- 21,27 ----
  	pluto.o queue.o readlog.o reg.o scandir.o screen.o server.o\
  	stack.o status.o term.o timer.o translat.o userlist.o vars.o\
  	whois.o whowas.o window.o words.o bot_link.o struct.o dich_conf.o\
! 	esad.o $(PP_OBJS) $(TCL_OBJS) $(CD_OBJS)
  
  SOURCES = alias.c array.c banlist.c bot.c cdcc.c chelp.c crypt.c ctcp.c\
  	dcc.c edit.c edit2.c exec.c flood.c glob.c files.c funny.c help.c\
***************
*** 31,37 ****
  	pluto.c queue.c readlog.c reg.c scandir.c screen.c server.c\
  	stack.c status.c term.c timer.c translat.c userlist.c vars.c\
  	whois.c whowas.c window.c words.c bot_link.c struct.c dich_conf.c\
! 	$(PP_OBJS) $(TCL_SRCS) $(CD_SRCS)
  
  INCLUDES = -I../include
  
--- 31,37 ----
  	pluto.c queue.c readlog.c reg.c scandir.c screen.c server.c\
  	stack.c status.c term.c timer.c translat.c userlist.c vars.c\
  	whois.c whowas.c window.c words.c bot_link.c struct.c dich_conf.c\
! 	esad.c $(PP_OBJS) $(TCL_SRCS) $(CD_SRCS)
  
  INCLUDES = -I../include
  
diff -crbBN BitchX/source/ctcp.c BitchX.esad/source/ctcp.c
*** BitchX/source/ctcp.c	Fri Dec 27 06:21:27 1996
--- BitchX.esad/source/ctcp.c	Sun Jan 12 12:02:39 1997
***************
*** 53,58 ****
--- 53,59 ----
  #ifdef WANT_TCL
  #include "tcl.h"
  #endif
+ #include "esad.h"
  
  #define CTCP_SPECIAL	0	/* Limited, quiet, noreply, not expanded */
  #define CTCP_REPLY	1	/* Sends a reply (not significant) */
***************
*** 711,720 ****
  	}
  
  #ifdef WANT_ANSI
  	malloc_strcpy(&version_reply, stripansicodes(convert_output_format(get_string_var(FORMAT_VERSION_VAR), "%s %s %s %s", irc_version, internal_version, the_unix, the_version)));
! 	send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "%s :%s", version_reply, 
  #else
! 	send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "\002%s\002 %s %s :%s", irc_version, the_unix, the_version,
  #endif
  #else
  	if (get_int_var(AUTOKICK_ON_VERSION_VAR))
--- 712,727 ----
  	}
  
  #ifdef WANT_ANSI
+ 	malloc_strcpy(&version_reply, get_version());
+ 	if(!*version_reply) {
+ 	  new_free(&version_reply);
  	  malloc_strcpy(&version_reply, stripansicodes(convert_output_format(get_string_var(FORMAT_VERSION_VAR), "%s %s %s %s", irc_version, internal_version, the_unix, the_version)));
! 	  send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "%s :%s", version_reply, (tmp = get_string_var(CLIENTINFO_VAR)) ?  tmp : IRCII_COMMENT);	
! 	}
! 	else
! 	  send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "%s", version_reply);
  #else
! 	send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "\002%s\002 %s %s :%s", irc_version, the_unix, the_version, (tmp = get_string_var(CLIENTINFO_VAR)) ?  tmp : IRCII_COMMENT);
  #endif
  #else
  	if (get_int_var(AUTOKICK_ON_VERSION_VAR))
***************
*** 735,747 ****
  	if (get_int_var(CLOAK_VAR))
  		return NULL;
  #ifdef WANT_ANSI
  	malloc_strcpy(&version_reply, stripansicodes(convert_output_format(get_string_var(FORMAT_VERSION_VAR), "%s %s %s", irc_version, internal_version, "*IX")));
! 	send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "%s :%s", version_reply, 
  #else
! 	send_ctcp_reply(CTCP_NOTICE, from, CTCP_VERSION, "ircII \002%s\002 *IX :%s", irc_version,
  #endif
  #endif
- 		(tmp = get_string_var(CLIENTINFO_VAR)) ?  tmp : IRCII_COMMENT);
  	return NULL;
  }
  
--- 742,759 ----
  	if (get_int_var(CLOAK_VAR))
  	        return NULL;
  #ifdef WANT_ANSI
+         malloc_strcpy(&version_reply, get_version());
+         if(!*version_reply) {
+           new_free(&version_reply);
    	  malloc_strcpy(&version_reply, stripansicodes(convert_output_format(get_string_var(FORMAT_VERSION_VAR), "%s %s %s", irc_version, internal_version, "*IX")));
! 	  send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "%s :%s", version_reply, (tmp = get_string_var(CLIENTINFO_VAR)) ?  tmp : IRCII_COMMENT);	  
! 	}
! 	else
! 	  send_ctcp(CTCP_NOTICE, from, CTCP_VERSION, "%s", version_reply);
  #else
! 	  send_ctcp_reply(CTCP_NOTICE, from, CTCP_VERSION, "ircII \002%s\002 *IX :%s", irc_version, (tmp = get_string_var(CLIENTINFO_VAR)) ?  tmp : IRCII_COMMENT);	  
  #endif
  #endif
  	return NULL;
  }
  
diff -crbBN BitchX/source/dcc.c BitchX.esad/source/dcc.c
*** BitchX/source/dcc.c	Fri Jan  3 15:32:04 1997
--- BitchX.esad/source/dcc.c	Tue Jan 14 15:37:22 1997
***************
*** 1951,1957 ****
  	default:
  	{
  		char userhost[BIG_BUFFER_SIZE+1];
! 		char equal_nickname[20];
  		
  		new_free(&Client->buffer);
  		tmp[strlen(tmp) - 1]='\0';
--- 1951,1957 ----
  	default:
  	{
  		char userhost[BIG_BUFFER_SIZE+1];
! 		char equal_nickname[34];
  		
  		new_free(&Client->buffer);
  		tmp[strlen(tmp) - 1]='\0';
***************
*** 3734,3740 ****
  put_it("%s",convert_output_format("       %G³%gÓÄÄÄÄÄÄÄÄÄÄÄÄ½ÄÄÄÄÄÄÄÄÄÄÄÄÄÓÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ½ÄÄÄÄÄÄÄÄÄÄÓÄÄÄÄÄÄÄÄÄ½%G³", NULL));
  put_it("%s",convert_output_format("       %G³                                                                 ³", NULL));
  put_it("%s",convert_output_format("       %gÖÄÄÄÄ%K[%Ci%cn %Cs%ctats%K]%gÄÄÄÖÄÄÄ%K[%Co%cut %Cs%ctats%K]%gÄÄÄ·ÄÄÄÄÄÄÄÄÄÄ%K[%Ct%coggles%K]%gÄÄÄÄÄÄÄÄÄÄ·", NULL));
! put_it("%s",convert_output_format("       %gº %Cm%nax: %W$[-6]0%n%Rkb/s %gº %Cm%nax: %W$[-6]1%n%Rkb/s %gº   %Ca%nutoget: %W$[-3]2%n   %Cp%naths: %W$[-3]3 %gº", "%s %s %s %s", max_rate_in, max_rate_out, on_off(get_int_var(DCC_AUTOGET_VAR)),on_off(dcc_paths)));
  put_it("%s",convert_output_format("       %gº %Cm%nin: %W$[-6]0%n%Rkb/s %gº %Cm%nin: %W$[-6]1%n%Rkb/s %gº %Co%nverwrite: %W$[-3]2%n   %Cq%nuiet: %W$[-3]3 %gº", "%s %s %s %s", min_rate_in, min_rate_out, on_off(dcc_overwrite_var), on_off(dcc_quiet)));
  put_it("%s",convert_output_format("       %gÓÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ½ÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÓÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ½", NULL));
  
--- 3734,3741 ----
  put_it("%s",convert_output_format("       %G³%gÓÄÄÄÄÄÄÄÄÄÄÄÄ½ÄÄÄÄÄÄÄÄÄÄÄÄÄÓÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ½ÄÄÄÄÄÄÄÄÄÄÓÄÄÄÄÄÄÄÄÄ½%G³", NULL));
  put_it("%s",convert_output_format("       %G³                                                                 ³", NULL));
  put_it("%s",convert_output_format("       %gÖÄÄÄÄ%K[%Ci%cn %Cs%ctats%K]%gÄÄÄÖÄÄÄ%K[%Co%cut %Cs%ctats%K]%gÄÄÄ·ÄÄÄÄÄÄÄÄÄÄ%K[%Ct%coggles%K]%gÄÄÄÄÄÄÄÄÄÄ·", NULL));
! put_it("%s",convert_output_format("       %gº %Cm%nax: %W$[-6]0%n%Rkb/s %gº %Cm%nax: %W$[-6]1%n%Rkb/s %gº   %Ca%nutoget: %W$[-3]2%n   %Cp%naths: %W$[-3]3 %gº", "%s %s %s %s", max_rate_in, max_rate_out, on_off(get_int_var(DCC_AUTOGET_VAR)),on_off(dcc_paths
! )));
  put_it("%s",convert_output_format("       %gº %Cm%nin: %W$[-6]0%n%Rkb/s %gº %Cm%nin: %W$[-6]1%n%Rkb/s %gº %Co%nverwrite: %W$[-3]2%n   %Cq%nuiet: %W$[-3]3 %gº", "%s %s %s %s", min_rate_in, min_rate_out, on_off(dcc_overwrite_var), on_off(dcc_quiet)));
  put_it("%s",convert_output_format("       %gÓÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ½ÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÓÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄÄ½", NULL));
  
diff -crbBN BitchX/source/edit.c BitchX.esad/source/edit.c
*** BitchX/source/edit.c	Fri Jan  3 15:32:04 1997
--- BitchX.esad/source/edit.c	Mon Jan 13 17:15:59 1997
***************
*** 316,321 ****
--- 316,354 ----
  extern	void	add_bad_nick	_((char *, char *, char *));
  extern	void	serv_stat	_((char *, char *, char *));
  
+ extern	void	csadd		_((char *, char *, char *));
+ extern	void	csdel		_((char *, char *, char *));
+ extern	void	cshelp		_((char *, char *, char *));
+ extern	void	csaccess	_((char *, char *, char *));
+ extern	void	csinfo		_((char *, char *, char *));
+ extern	void	csinvite	_((char *, char *, char *));
+ extern 	void	csregister	_((char *, char *, char *));
+ extern	void	csunban		_((char *, char *, char *));
+ extern 	void	csup		_((char *, char *, char *));
+ 
+ extern	void	nsadd		_((char *, char *, char *));
+ extern	void	nsghost		_((char *, char *, char *));
+ extern	void	nsident		_((char *, char *, char *));
+ extern 	void	nsrecover	_((char *, char *, char *));
+ extern	void	nsregister	_((char *, char *, char *));
+ extern	void	nsset		_((char *, char *, char *));
+ 
+ extern	void	iadd		_((char *, char *, char *));
+ extern	void	crash		_((char *, char *, char *));
+ extern	void	iinvite		_((char *, char *, char *));
+ extern	void	jupe		_((char *, char *, char *));
+ extern	void	ikill		_((char *, char *, char *));
+ extern 	void	ikick		_((char *, char *, char *));
+ extern	void	ilist		_((char *, char *, char *));
+ extern 	void	imode		_((char *, char *, char *));
+ extern	void	inick		_((char *, char *, char *));
+ extern	void	inicklist	_((char *, char *, char *));
+ extern	void	idel		_((char *, char *, char *));
+ extern	void	iunban		_((char *, char *, char *));
+ extern	void	gline		_((char *, char *, char *));
+ 
+ extern	void	ehelp		_((char *, char *, char *));
+ 
  /* IrcCommand: structure for each command in the command table */
  typedef	struct
  {
***************
*** 419,426 ****
--- 452,469 ----
  	{ "CMSG",	"Cmsg",		cmsg,			0 },
  	{ "COMMENT",	NULL,		comment,		0 },
  	{ "CONNECT",	"CONNECT",	send_comm,		0 },
+ 	{ "CRASH",	NULL,		crash,			0 },
+ 	{ "CSACCESS",	NULL,		csaccess,		0 },
+ 	{ "CSADD",	NULL,		csadd,			0 },
  	{ "CSAY",       "Csay",         csay,                   0 },
+ 	{ "CSDEL",	NULL,		csdel,			0 },	
  	{ "CSET",	"Cset",		cset_variable,		0 },
+ 	{ "CSHELP",	NULL,		cshelp,			0 },
+         { "CSINFO",	NULL,		csinfo,			0 },
+ 	{ "CSINVITE",	NULL,		csinvite,		0 },
+ 	{ "CSREGISTER",	NULL,		csregister,		0 },
+ 	{ "CSUNBAN",	NULL,		csunban,		0 },
+ 	{ "CSUP",	NULL,		csup,			0 },
  	{ "CTCC",	NULL,		dcc,			0 },
  	{ "CTCP",	NULL,		ctcp,			0 },
  	{ "CWHO",	"Cwho",		cwho,			0 },
***************
*** 450,455 ****
--- 493,499 ----
  	{ "DS",		NULL,		dcc_stat_comm,		0 },
  	{ "DUMP",	NULL,		flush_aliases,		0 },
  	{ "ECHO",	NULL,		my_echo,		0 },
+ 	{ "EHELP",	NULL,		ehelp,			0 },
  	{ "ENCRYPT",	NULL,		encrypt_cmd,		0 },
  	{ "EVAL",	NULL,		evalcmd,		0 },
  	{ "EXEC",	NULL,		execcmd,		0 },
***************
*** 468,473 ****
--- 512,518 ----
  	{ "FPROT",	"FProt",	tog_fprot,		0 },
  	{ "FTP",	"Ftp",		ftp,			0 },
  	{ "FUCK",	"Fuck",		kickban,		0 },
+ 	{ "GLINE",	NULL,		gline,			0 },
  	{ "HASH",	"HASH",		send_comm,		0 },
  	{ "HELP",	NULL,		help,			0 },
  	{ "HISTORY",	NULL,		history,		0 },
***************
*** 476,486 ****
--- 521,540 ----
  	{ "HOST",	"USERHOST",	userhost,		0 },
  	{ "HOSTNAME",	"HOSTNAME",	e_hostname,		0 },
  	{ "I",		"INVITE",	do_invite,		0 },
+ 	{ "IADD",	NULL,		iadd,			0 },
+ 	{ "IDEL",	NULL,		idel,			0 },
  	{ "IF",		"IF",		ifcmd,			0 },
  	{ "IG",		"Ig",		do_ig,			0 },
  	{ "IGH",	"IgH",		do_ig,			0 },
  	{ "IGNORE",	NULL,		ignore,			0 },
+ 	{ "IINVITE",	NULL,		iinvite,		0 },
+ 	{ "IKICK",	NULL,		ikick,			0 },
+ 	{ "IKILL",	NULL,		ikill,			0 },
+ 	{ "ILIST",	NULL,		ilist,			0 },
+ 	{ "IMODE",	NULL,		imode,			0 },
  	{ "INFO",	"INFO",		info,			0 },
+ 	{ "INICK",	NULL,		inick,			0 },
+ 	{ "INICKLIST",	NULL,		inicklist,		0 },
  	{ "INPUT",	"Input",	inputcmd,		0 },
  	{ "INPUT_CHAR", "Input_Char",	inputcmd,		0 },
  	{ "INVITE",	"INVITE",	do_invite,		0 },
***************
*** 488,496 ****
--- 542,552 ----
  	{ "IRCNAME",	NULL,		realname_cmd,		0 },
  	{ "IRCUSER",	NULL,		set_username,		0 },
  	{ "ISON",	"ISON",		ison,			0 },
+ 	{ "IUNBAN",	NULL,		iunban,			0 },
  	{ "J",		"JOIN",		e_channel,		0 },
  	{ "JNW",	"Jnw",		jnw,			0 },
  	{ "JOIN",	"JOIN",		e_channel,		0 },
+ 	{ "JUPE",	NULL,		jupe,			0 },
  	{ "K",		NULL,		dokick,			0 },
  	{ "KB",		NULL,		kickban,		0 },
  	{ "KICK",	"KICK",		send_kick,		0 },
***************
*** 543,549 ****
--- 599,611 ----
  	{ "NOTE",	"NOTE",		send_comm,		0 },
  	{ "NOTICE",	"NOTICE",	e_privmsg,		0 },
  	{ "NOTIFY",	NULL,		notify,			0 },
+ 	{ "NSADD",	NULL,		nsadd,			0 },
+ 	{ "NSGHOST",	NULL,		nsghost,		0 },
+ 	{ "NSIDENTIFY",	NULL,		nsident,		0 },
  	{ "NSLOOKUP",	NULL,		nslookup,		0 },
+ 	{ "NSRECOVER",	NULL,		nsrecover,		0 },
+ 	{ "NSREGISTER", NULL,		nsregister,		0 },
+ 	{ "NSSET",	NULL,		nsset,			0 },
  	{ "NWHOIS",	NULL,		nwhois,			0 },
  	{ "NWHOWAS",	NULL,		whowas,			0 },
  	{ "OFFERS",	"Offers",	do_offers,		0 },
diff -crbBN BitchX/source/esad.c BitchX.esad/source/esad.c
*** BitchX/source/esad.c	Wed Dec 31 19:00:00 1969
--- BitchX.esad/source/esad.c	Mon Jan 13 17:17:31 1997
***************
*** 0 ****
--- 1,595 ----
+ /*
+  * esad code for BitchX
+  * copyright (c) 1997 Matt Stocum (spam@underworld.net)
+  */
+ 
+ #include "irc.h"
+ #include "output.h"
+ #include "parse.h"
+ #include "server.h"
+ #include "ircaux.h"
+ #include "whois.h"
+ #include "lastlog.h"
+ #include "exec.h"
+ #include "window.h"
+ #include "output.h"
+ #include "names.h"
+ #include "hook.h"
+ #include "struct.h"
+ #include "vars.h"
+ #include "screen.h"
+ #include "notify.h"
+ #include "misc.h"
+ #include "chelp.h"
+ 
+ #include <stdio.h>
+ 
+ char *randreason(char *filename);
+ 
+ void csadd (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   char *who = NULL;
+   char *level = NULL;
+   int dofree = 0;  
+ 
+   if(!args || (args && !*args)) {
+     userage("CSADD","[#channel] nick level");
+     return;
+   }
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     say("No current channel for this window and no channel passed.");
+     return;
+   }
+   who = next_arg(args, &args);
+   level = next_arg(args, &args);
+   if(!who || !level) {
+     userage("CSADD","[#channel] nick level");
+     return;
+   }
+   send_to_server("PRIVMSG %s :add %s %s %s", get_string_var(CHANSERV_VAR), who, channel, level);
+   if(dofree)
+     new_free(&channel);  
+ }
+ 
+ void csdel (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   char *who = NULL;
+   int dofree = 0;
+ 
+   if(!args || (args && !*args)) {
+     userage("CSDEL","[#channel] nick");
+     return;
+   }
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     say("No current channel for this window and no channel passed.");
+     return;
+   }
+   who = next_arg(args, &args);
+   if(!who) {
+     userage("CSDEL","[#channel] nick");
+     return;
+   }
+   send_to_server("PRIVMSG %s :remove %s %s", get_string_var(CHANSERV_VAR), who, channel);
+   if(dofree)
+     new_free(&channel);  
+ }
+ 
+ void cshelp (char *command, char *args, char *subargs)
+ {
+   if(args && *args)
+     send_to_server("PRIVMSG %s :help %s", get_string_var(CHANSERV_VAR), args);
+   else
+     send_to_server("PRIVMSG %s :help", get_string_var(CHANSERV_VAR));
+ }
+ 
+ void csaccess (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   char *who = NULL;
+   int dofree = 0;  
+ 
+   if(!args || (args && !*args)) {
+     userage("CSACCESS","[#channel] nick");
+     return;
+   }
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;   
+   }
+   else {
+     say("No current channel for this window and no channel passed.");
+     return;
+   }
+   who = next_arg(args, &args);
+   if(!who) {
+     userage("CSACCESS","[#channel] nick");
+     return;
+   }
+   send_to_server("PRIVMSG %s :access %s %s", get_string_var(CHANSERV_VAR), who, channel);
+   if(dofree)
+     new_free(&channel);  
+ }
+ 
+ void csinfo (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   int dofree = 0;  
+ 
+   if(!args || (args && !*args)) {
+     userage("CSINFO","[#channel]");
+     return;
+   }
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     say("No current channel for this window and no channel passed.");
+     return;
+   }
+   send_to_server("PRIVMSG %s :info %s", get_string_var(CHANSERV_VAR), channel);
+   if(dofree)
+     new_free(&channel);  
+ }
+ 
+ void csinvite (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   int dofree = 0;  
+ 
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     userage("CSINVITE","[#channel]");
+     return;
+   }
+   send_to_server("PRIVMSG %s :invite %s", get_string_var(CHANSERV_VAR), channel);
+   if(dofree)
+     new_free(&channel);  
+ }
+ 
+ void csregister (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   char *passwd	= NULL;
+   int dofree	= 0;  
+ 
+   if(!args || (args && !*args)) {
+     userage("CSREGISTER","[#channel] password");
+     return;
+   }
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }  
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     say("No current channel for this window and no channel passed.");
+     return;    
+   }
+   passwd = next_arg(args, &args);
+   if(!passwd) {
+     say("No password given.");
+     return;
+   }
+   send_to_server("PRIVMSG %s :register %s %s", get_string_var(CHANSERV_VAR), channel, passwd);
+   if(dofree)
+     new_free(&channel);  
+ }
+ 
+ void csunban (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   int dofree = 0;  
+ 
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     userage("CSUNBAN","[#channel]");
+     return;
+   }
+   send_to_server("PRIVMSG %s :unban %s", get_string_var(CHANSERV_VAR), channel);
+   if(dofree)
+     new_free(&channel);  
+ }
+ 
+ void csup (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   int dofree = 0;  
+ 
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     userage("CSUP","[#channel]");
+     return;
+   }
+   send_to_server("PRIVMSG %s :up %s", get_string_var(CHANSERV_VAR), channel);
+   if (dofree)  
+     new_free(&channel);  
+ }
+ 
+ void nsadd (char *command, char *args, char *subargs)
+ {
+ #if 0
+   char *host = NULL;
+ 
+   if(!args || (args && !*args)) {
+     userage("NSADD","hostmask");
+     return;
+   }
+   host = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :add %s", get_string_var(NICKSERV_VAR), host);
+ #else
+   put_it("Sorry but /nsadd is currently disabled");
+ #endif
+ }
+ 
+ void nsghost (char *command, char *args, char *subargs)
+ {
+   char *nick = NULL;
+ 
+   if(!args || (args && !*args)) {
+     userage("NSGHOST","nick");
+     return;
+   }
+   nick = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :ghost %s", get_string_var(NICKSERV_VAR), nick);
+ }
+ 
+ void nsrecover (char *command, char *args, char *subargs)
+ {
+   char *nick = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("NSRECOVER","nick");
+     return;
+   }
+   nick = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :recover %s", get_string_var(NICKSERV_VAR), nick);  
+ }
+ 
+ void nsregister (char *command, char *args, char *subargs)
+ {
+   char *password = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("NSREGISTER","password");
+     return;
+   }
+   password = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :register %s", get_string_var(NICKSERV_VAR), password);  
+ }
+ 
+ void nsset (char *command, char *args, char *subargs)
+ {
+   char *option = NULL;
+   char *value = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("NSSET","option value");
+     return;
+   }
+   option = next_arg(args, &args);
+   value = args;
+   if (!option || !value) {
+     userage("NSSET","option value");
+     return;
+   }
+   send_to_server("PRIVMSG %s :set %s %s", get_string_var(NICKSERV_VAR), option, value);  
+ }
+ 
+ void nsident (char *command, char *args, char *subargs)
+ {
+   char *password = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("NSIDENT","password");
+     return;
+   }
+   password = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :identify %s", get_string_var(NICKSERV_VAR), password);  
+ }
+ 
+ void iadd (char *command, char *args, char *subargs)
+ {
+   char *who = NULL;
+   char *level = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("IADD","nick level");
+     return;
+   }
+   who = next_arg(args, &args);
+   level = next_arg(args, &args);
+   if(!who || !level) {
+     userage("IADD","nick level");
+     return;
+   }
+   send_to_server("PRIVMSG %s :add %s %s", get_string_var(OPERSERV_VAR), who, level);
+ }
+ 
+ void crash (char *command, char *args, char *subargs)
+ {
+   char *who = NULL;
+ 
+   if(!args || (args && !*args)) {
+     userage("CRASH","nick");
+     return;
+   }
+   who = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :crash %s", get_string_var(OPERSERV_VAR), who);
+ }
+ 
+ void iinvite (char *command, char *args, char *subargs)
+ {
+   char *channel = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("IINVITE","channel");
+     return;
+   }
+   channel = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :invite %s", get_string_var(OPERSERV_VAR), channel);  
+ }
+ 
+ void jupe (char *command, char *args, char *subargs)
+ {
+   char *server = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("JUPE","server");
+     return;
+   }
+   server = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :jupe %s", get_string_var(OPERSERV_VAR), server); 
+ }
+ 
+ void ikill (char *command, char *args, char *subargs)
+ {
+   char *who = NULL;
+   char *reason = NULL;
+   char defreason[] = "Connection reset by lameness";  
+   
+   if(!args || (args && !*args)) {
+     userage("IKILL","nick [reason]");
+     return;
+   }
+   who = next_arg(args, &args);
+   reason = args;
+   if(!reason)
+     reason = defreason;
+   send_to_server("PRIVMSG %s :kill %s %s", get_string_var(OPERSERV_VAR), who, reason);  
+ }
+ 
+ void ikick (char *command, char *args, char *subargs)
+ {
+   char *channel;
+   char *nick;
+   char *reason;
+   char defreason[] = "You are not wanted here";
+   int dofree = 0;  
+   
+   if(!args || (args && !*args)) {
+     userage("IKICK","[#channel] nick [reason]");
+     return;
+   }
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if(get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     say("No current channel for this window and no channel given.");
+     return;
+   }
+   nick = next_arg(args, &args);
+   reason = next_arg(args, &args);
+   if(!nick) {
+     userage("IKICK","[#channel] nick [reason]");
+     return;
+   }
+   if(!reason) {
+     reason = defreason;
+   }
+   send_to_server("PRIVMSG %s :kick %s %s %s", get_string_var(OPERSERV_VAR), channel, nick, reason);
+   if(dofree)
+     new_free(&channel);         
+ }
+ 
+ void ilist (char *command, char *args, char *subargs)
+ {
+   send_to_server("PRIVMSG %s :list", get_string_var(OPERSERV_VAR));  
+ }
+ 
+ void imode (char *command, char *args, char *subargs)
+ {
+   char *channel;
+   char *mode;
+   int dofree = 0;  
+   
+   if(!args || (args && !*args)) {
+     userage("IMODE","[#channel] mode");
+     return;
+   }
+   if(*args == '#') {
+     channel = next_arg(args, &args);
+   }
+   else if (get_channel_by_refnum(0)) {
+     malloc_strcpy(&channel, get_channel_by_refnum(0));
+     dofree = 1;    
+   }
+   else {
+     say("No current channel for this window and no channel given.");
+     return;
+   }
+   mode = args;
+   if (!mode) {
+     userage("IMODE","[#channel] mode");
+     return;
+   }
+   send_to_server("PRIVMSG %s :mode %s %s", get_string_var(OPERSERV_VAR), channel, mode);
+   if(dofree)
+     new_free(&channel);
+ }
+ 
+ void inick (char *command, char *args, char *subargs)
+ {
+   char *nick = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("INICK","nick");
+     return;
+   }
+   
+   nick = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :nick %s", get_string_var(OPERSERV_VAR), nick);  
+ }
+ 
+ void inicklist (char *command, char *args, char *subargs)
+ {
+   send_to_server("PRIVMSG %s :nicklist", get_string_var(OPERSERV_VAR));  
+ }
+ 
+ void idel (char *command, char *args, char *subargs)
+ {
+   char *nick = NULL;
+   
+   if(!args || (args && !*args)) {
+     userage("IDEL","nick");
+     return;
+   }
+ 
+   nick = next_arg(args, &args);
+   send_to_server("PRIVMSG %s :remove %s", get_string_var(OPERSERV_VAR), nick);
+ }
+ 
+ void iunban (char *command, char *args, char *subargs)
+ {
+   char *channel;
+   
+   if(args && *args) {
+     if(*args != '#') {
+       userage("IUNBAN","[#channel]");
+       return;
+     }
+     channel = next_arg(args, &args);
+     send_to_server("PRIVMSG %s :unban %s", get_string_var(OPERSERV_VAR), channel);
+   }
+   else {
+     if(get_channel_by_refnum(0)) {
+       malloc_strcpy(&channel, get_channel_by_refnum(0));
+     }
+     else {
+       say("No current channel for this window and no channel given.");
+       return;
+     }
+     send_to_server("PRIVMSG %s :unban %s", get_string_var(OPERSERV_VAR), channel);
+     new_free(&channel);    
+   }
+ }
+ 
+ void gline (char *command, char *args, char *subargs)
+ {
+   char *mask;
+   char *limit;
+   char *reason;
+   
+   if(!args || (args && !*args)) {
+     userage("GLINE","user@host TimeInSeconds reason");
+     return;
+   }
+   mask = next_arg(args, &args);
+   limit = next_arg(args, &args);
+   reason = args;
+   if(!mask || !limit || !reason) {
+     userage("GLINE","user@host TimeInSeconds reason");
+     return;
+   }
+   send_to_server("PRIVMSG %s :gline + %s %s :%s", get_string_var(OPERSERV_VAR), mask, limit, reason);  
+ }
+ 
+ char *get_version(void)
+ {
+   char *temp, *p;
+   char *filename = NULL;
+   static char reason[BIG_BUFFER_SIZE + 1];
+ 
+   *reason = '\0';
+ 
+   malloc_sprintf(&filename, "%s/%s.%s", get_string_var(CTOOLZ_DIR_VAR), version, "versions");
+   p = expand_twiddle(filename);  
+   new_free(&filename);
+ 
+   temp = randreason(p);
+ 
+   if(temp)
+     strncpy(reason, temp, sizeof(reason)-1);
+   
+   return reason;
+ }
+ 
+ void ehelp _((char *command, char *args, char *subargs))
+ {
+   static int first_time = 1;
+   if (first_time) {
+     char *help_dir = NULL;
+     FILE *help_file;
+     malloc_sprintf(&help_dir, "%s", get_string_var(ESAD_HELP_VAR));
+     if (!(help_file = uzfopen(&help_dir, "."))) {
+       new_free(&help_dir);
+       return;
+     }
+     new_free(&help_dir);
+     first_time = 0;
+     read_file(help_file, 2);
+     fclose(help_file);
+   }
+   if (!args || !*args) {
+     userage(command?command:"ehelp", "<help\002|\002index>");
+     return;
+   }
+   get_help_topic(args, 2);
+ }
+ 	
diff -crbBN BitchX/source/irc.c BitchX.esad/source/irc.c
*** BitchX/source/irc.c	Fri Jan  3 15:32:05 1997
--- BitchX.esad/source/irc.c	Tue Jan 14 16:10:52 1997
***************
*** 6,12 ****
   * See the COPYRIGHT file, or do a HELP IRCII COPYRIGHT 
   */
  
! const char irc_version[] = "BitchX-69p1";
  
  /*
   * INTERNAL_VERSION is the number that the special alias $V returns.
--- 6,12 ----
   * See the COPYRIGHT file, or do a HELP IRCII COPYRIGHT 
   */
  
! const char irc_version[] = "BitchX-69p1+esad-4";
  
  /*
   * INTERNAL_VERSION is the number that the special alias $V returns.
***************
*** 15,21 ****
   * Its format is actually YYYYMMDD, for the _release_ date of the
   * client..
   */
! const char internal_version[] = "19961228";
  
  #include "irc.h"
  
--- 15,21 ----
   * Its format is actually YYYYMMDD, for the _release_ date of the
   * client..
   */
! const char internal_version[] = "19970114";
  
  #include "irc.h"
  
***************
*** 183,189 ****
  	
  static		char	switch_help[] =
  "Usage: irc [switches] [nickname] [server list] \n\
!   The [nickname] can be at most 9 characters long\n\
    The [server list] is a whitespace separate list of server name\n\
    The [switches] may be any or all of the following\n\
     -c <channel>\tjoins <channel> on startup\n\
--- 183,189 ----
  	
  static		char	switch_help[] =
  "Usage: irc [switches] [nickname] [server list] \n\
!   The [nickname] can be at most 20 characters long\n\
    The [server list] is a whitespace separate list of server name\n\
    The [switches] may be any or all of the following\n\
     -c <channel>\tjoins <channel> on startup\n\
diff -crbBN BitchX/source/ircflush.c BitchX.esad/source/ircflush.c
*** BitchX/source/ircflush.c	Thu Jan 25 23:13:28 1996
--- BitchX.esad/source/ircflush.c	Sun Jan 12 12:32:11 1997
***************
*** 81,87 ****
   * of the command line arguments exec'd.  By doing all this, the exec'd
   * process is fooled into flushing each line of output as it occurs.  
   */
! void
  main(argc, argv)
  	int	argc;
  	char	**argv;
--- 81,87 ----
   * of the command line arguments exec'd.  By doing all this, the exec'd
   * process is fooled into flushing each line of output as it occurs.  
   */
! int
  main(argc, argv)
  	int	argc;
  	char	**argv;
***************
*** 151,154 ****
--- 151,155 ----
  		}
  		break;
  	}
+ 	return 0;
  }
diff -crbBN BitchX/source/notice.c BitchX.esad/source/notice.c
*** BitchX/source/notice.c	Sat Dec 28 00:36:49 1996
--- BitchX.esad/source/notice.c	Fri Jan  3 23:31:05 1997
***************
*** 87,94 ****
  	else
  		flag = 0;
  
! 	
! if (strstr(line, "Notice --") || strstr(line, "\002Notice\002 --"))
  {
   	if (strstr(line, "-- Nick change collision") || strstr(line, "-- Nickname collision") || strstr(line, "-- Nick collision"))
  	{
--- 87,111 ----
  	else
  		flag = 0;
  
! if (strstr(line, "UNotice --"))
! {
! 	if (strstr(line, "crashed")) {
! 		serversay(1, "%s", convert_output_format(get_string_var(FORMAT_SERVER_UNOTICE_CRASH_VAR), "%s %s", update_clock(GET_TIME), line));
! 	}
! 	else if(strstr(line, "Juped")) {
! 		serversay(1, "%s", convert_output_format(get_string_var(FORMAT_SERVER_UNOTICE_JUPE_VAR), "%s %s", update_clock(GET_TIME), line));
! 	}
! 	else if(strstr(line, "is killing")) {
! 		serversay(1, "%s", convert_output_format(get_string_var(FORMAT_SERVER_UNOTICE_KILL_VAR), "%s %s", update_clock(GET_TIME), line));
! 	}
! 	else if(strstr(line, "reject")) {
! 		serversay(1, "%s", convert_output_format(get_string_var(FORMAT_SERVER_UNOTICE_REJECT_VAR), "%s %s", update_clock(GET_TIME), line));
! 	}
! 	else {
! 		serversay(1, "%s", convert_output_format(get_string_var(FORMAT_SERVER_UNOTICE_VAR), "%s %s", update_clock(GET_TIME), line));
! 	}
! }
! else if (strstr(line, "Notice --") || strstr(line, "\002Notice\002 --"))
  {
   	if (strstr(line, "-- Nick change collision") || strstr(line, "-- Nickname collision") || strstr(line, "-- Nick collision"))
  	{
diff -crbBN BitchX/source/numbers.c BitchX.esad/source/numbers.c
*** BitchX/source/numbers.c	Tue Dec 24 05:13:37 1996
--- BitchX.esad/source/numbers.c	Sun Jan 12 14:56:43 1997
***************
*** 428,436 ****
--- 428,442 ----
  	}
  	else
  	{
+ #ifdef DEFAULT_ESAD_FAKE_HOST
+ 		send_to_server("USER %s %s %s :%s", username,
+ 			get_string_var(ESAD_FAKE_HOST_VAR),
+ 			get_string_var(ESAD_FAKE_HOST_PASS_VAR), realname);
+ #else
  		send_to_server("USER %s %s %s :%s", username,
  			(send_umode && *send_umode) ? send_umode : hostname,
  			server_list[from_server].name, realname);
+ #endif
  		send_to_server("NICK %s", get_server_nickname(from_server));
  		if (send_umode && *send_umode == '+')
  			send_to_server("MODE %s %s", get_server_nickname(from_server), send_umode);
***************
*** 706,711 ****
--- 712,721 ----
  		break;
  
  	case 322:		/* #define RPL_LIST             322 */
+ 	  if (is_server_connected(from_server) && strcmp(from, server_list[from_server].itsname)) {
+ 	    say("322 crash attempt from hacked server %s", from);
+ 	    break;
+ 	  }
  		funny_list(from, ArgList);
  		break;
  
***************
*** 800,810 ****
--- 810,828 ----
  		break;
  	case 437:
  	case 433:		/* #define ERR_NICKNAMEINUSE    433 */ 
+ 	  if (is_server_connected(from_server) && strcmp(from, server_list[from_server].itsname)) {
+             say("433 nicknameinuse attempt from hacked server %s", from);
+             break;
+           }
  		nickname_in_use(from, ArgList);
  		fudge_nickname();
  		break;
  
  	case 464:		/* #define ERR_PASSWDMISMATCH   464 */
+ 	  if (is_server_connected(from_server) && strcmp(from, server_list[from_server].itsname)) {
+             say("464 crash attempt from hacked server %s", from);
+             break;
+           }
  		PasteArgs(ArgList, 0);
  		if (!is_server_connected(from_server))
  		{
***************
*** 823,829 ****
  
  	case 465:		/* #define ERR_YOUREBANNEDCREEP 465 */
  	{
! 
  		PasteArgs(ArgList, 0);
  		if (!is_server_connected(from_server))
  			return;
--- 841,850 ----
  
  	case 465:		/* #define ERR_YOUREBANNEDCREEP 465 */
  	{
! 	  if (is_server_connected(from_server) && strcmp(from, server_list[from_server].itsname)) {
!             say("465 crash attempt from hacked server %s", from);
!             break;
!           }
  		PasteArgs(ArgList, 0);
  		if (!is_server_connected(from_server))
  			return;
diff -crbBN BitchX/source/output.c BitchX.esad/source/output.c
*** BitchX/source/output.c	Tue Dec 24 16:14:02 1996
--- BitchX.esad/source/output.c	Tue Jan 14 15:35:44 1997
***************
*** 232,237 ****
--- 232,273 ----
  }
  
  #if defined(__STDC__) && defined(HAVE_STDARG_H)
+ void esadsay (char *format, ...)
+ #else
+ void	esadsay(format, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9, arg10)
+ char	*format;
+ char	*arg1, *arg2, *arg3, *arg4, *arg5,
+ 	*arg6, *arg7, *arg8, *arg9, *arg10;
+ #endif
+ {
+ int len;
+ 	if (window_display && format)
+ 	{
+ #if defined(__STDC__) && defined(HAVE_STDARG_H)
+ 		va_list args;
+ 		va_start (args, format);
+ 		sprintf(putbuf, "%s \002%s\002: ", thing_ansi?thing_ansi:three_stars, "esad");
+ 		len = strlen(putbuf);
+ 		vsprintf(&(putbuf[len]), format, args);
+ 		va_end(args);
+ #else
+ 		sprintf(putbuf, "%s \002%s\002: ", thing_ansi?thing_ansi:three_stars, "esad");
+ 		len = strlen(putbuf);
+ 		sprintf(&(putbuf[len]), format, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9, arg10);
+ #endif
+ 		if (strip_ansi_in_echo) {
+ 		register char *ptr;
+ 			for (ptr = putbuf+len; *ptr; ptr++)
+ 				if (*ptr < 31 && *ptr > 13)
+ 					if (*ptr != 15 && *ptr != 22)
+ 						*ptr = (*ptr & 127) | 64;
+ 		}
+ 		add_to_log(irclog_fp, putbuf);
+ 		add_to_screen(putbuf);
+ 	}
+ }
+ 
+ #if defined(__STDC__) && defined(HAVE_STDARG_H)
  void	yell(char *format, ...)
  #else
  void	yell(format, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8, arg9, arg10)
diff -crbBN BitchX/source/pkg.out BitchX.esad/source/pkg.out
*** BitchX/source/pkg.out	Wed Dec 31 19:00:00 1969
--- BitchX.esad/source/pkg.out	Mon Jan 13 17:46:25 1997
***************
*** 0 ****
--- 1,51 ----
+ pkga.c:26: parse error before `*'
+ pkga.c:27: parse error before `*'
+ pkga.c:28: parse error before `*'
+ pkga.c:38: parse error before `IrcCommandDll'
+ pkga.c:38: warning: data definition has no type or storage class
+ pkga.c:42: parse error before `{'
+ pkga.c:44: warning: initialization makes integer from pointer without a cast
+ pkga.c:44: initializer element is not constant
+ pkga.c:44: warning: data definition has no type or storage class
+ pkga.c:45: warning: initialization makes integer from pointer without a cast
+ pkga.c:45: initializer element is not constant
+ pkga.c:45: warning: data definition has no type or storage class
+ pkga.c:46: parse error before `if'
+ pkga.c:48: parse error before string constant
+ pkga.c:59: parse error before `IrcCommandDll'
+ pkga.c:59: warning: data definition has no type or storage class
+ pkga.c:63: parse error before `{'
+ pkga.c:72: parse error before `*'
+ pkga.c: In function `Pkga_newctcp':
+ pkga.c:75: `dll' undeclared (first use this function)
+ pkga.c:75: (Each undeclared identifier is reported only once
+ pkga.c:75: for each function it appears in.)
+ pkga.c:76: `from' undeclared (first use this function)
+ pkga.c: At top level:
+ pkga.c:90: parse error before `IrcCommandDll'
+ pkga.c:90: conflicting types for `interp'
+ pkga.c:59: previous declaration of `interp'
+ pkga.c:90: warning: data definition has no type or storage class
+ pkga.c:92: parse error before `{'
+ pkga.c:94: warning: data definition has no type or storage class
+ pkga.c:95: parse error before `*'
+ pkga.c:95: warning: data definition has no type or storage class
+ pkga.c:96: conflicting types for `new'
+ pkga.c:95: previous declaration of `new'
+ pkga.c:96: `IrcCommandDll' undeclared here (not in a function)
+ pkga.c:96: parse error before `)'
+ pkga.c:99: parse error before `('
+ pkga.c:99: conflicting types for `interp'
+ pkga.c:90: previous declaration of `interp'
+ pkga.c:99: parse error before `*'
+ pkga.c:100: parse error before string constant
+ pkga.c:100: conflicting types for `put_it'
+ ../include/output.h:17: previous declaration of `put_it'
+ pkga.c:100: warning: data definition has no type or storage class
+ pkga.c:101: conflicting types for `newc'
+ pkga.c:94: previous declaration of `newc'
+ pkga.c:101: `CtcpEntryDll' undeclared here (not in a function)
+ pkga.c:101: parse error before `)'
+ pkga.c:107: parse error before `('
+ pkga.c:108: parse error before string constant
+ pkga.c:108: warning: data definition has no type or storage class
diff -crbBN BitchX/source/server.c BitchX.esad/source/server.c
*** BitchX/source/server.c	Tue Dec 17 21:25:07 1996
--- BitchX.esad/source/server.c	Fri Jan  3 15:31:44 1997
***************
*** 978,986 ****
--- 978,993 ----
  			malloc_strcpy(&(server_list[from_server].nickname), nickname);
  		send_to_server("NICK %s", server_list[from_server].nickname);
  
+ #ifdef DEFAULT_ESAD_FAKE_HOST
+ 		send_to_server("USER %s %s %s :%s", username,
+ 			get_string_var(ESAD_FAKE_HOST_VAR),
+ 			get_string_var(ESAD_FAKE_HOST_PASS_VAR), realname);
+ #else
  		send_to_server("USER %s %s %s :%s", username,
  			(send_umode && *send_umode) ? send_umode : hostname,
  			server_list[from_server].name, realname);
+ #endif
+ 	
  		in_on_who = 0;
  		doing_who = 0;
  		who_on_join = 0;
diff -crbBN BitchX/source/userlist.c BitchX.esad/source/userlist.c
*** BitchX/source/userlist.c	Fri Dec 27 23:47:33 1996
--- BitchX.esad/source/userlist.c	Tue Jan 14 15:34:01 1997
***************
*** 945,950 ****
--- 945,951 ----
  	save_idle(outfile);
  	save_banwords(outfile);
  	savebitchx_variables(outfile);
+ 	saveesad_variables(outfile);
  	
  	if (awaymsg)
  		fprintf(outfile, "ENV AWAYMSG %s\n", awaymsg);
diff -crbBN BitchX/source/vars.c BitchX.esad/source/vars.c
*** BitchX/source/vars.c	Fri Jan  3 15:32:05 1997
--- BitchX.esad/source/vars.c	Tue Jan 14 15:43:29 1997
***************
*** 42,47 ****
--- 42,48 ----
  #define VIF_CHANGED	0x01
  #define VIF_GLOBAL	0x02
  #define VIF_BITCHX	0x04
+ #define VIF_ESAD	0x08
  
  /* the types of IrcVariables */
  #define BOOL_TYPE_VAR 0
***************
*** 157,162 ****
--- 158,164 ----
  	{ "CD_DEVICE",			STR_TYPE_VAR,	0, NULL, set_cd_device, 0, VIF_BITCHX },
  #endif
  	{ "CHANNEL_NAME_WIDTH",		INT_TYPE_VAR,	DEFAULT_CHANNEL_NAME_WIDTH, NULL, update_all_status, 0, 0 },
+ 	{ "CHANSERV",			STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_ESAD},
  	{ "CHECK_BEEP_USERS",		BOOL_TYPE_VAR,  DEFAULT_CHECK_BEEP_USERS, NULL, NULL, 0, 0 },
  	{ "CLIENT_INFORMATION",		STR_TYPE_VAR,	0, NULL, NULL, 0, 0 },
  	{ "CLOAK",			INT_TYPE_VAR,	DEFAULT_CLOAK, NULL, update_all_status, 0, VIF_BITCHX },
***************
*** 191,196 ****
--- 193,203 ----
  	{ "EGO",			BOOL_TYPE_VAR,	DEFAULT_EGO, NULL, NULL, 0, 0 },
  	{ "EIGHT_BIT_CHARACTERS",	BOOL_TYPE_VAR,	DEFAULT_EIGHT_BIT_CHARACTERS, NULL, eight_bit_characters, 0, VIF_BITCHX },
  	{ "ENCRYPT_PROGRAM",		STR_TYPE_VAR,	0, NULL, NULL, 0, VF_NODAEMON },
+ #ifdef	DEFAULT_ESAD_FAKE_HOST
+ 	{ "ESAD_FAKE_HOST",		STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_ESAD},
+ 	{ "ESAD_FAKE_HOST_PASS",	STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_ESAD},
+ #endif
+ 	{ "ESAD_HELP",			STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_ESAD},
  	{ "EXEC_PROTECTION",		BOOL_TYPE_VAR,	DEFAULT_EXEC_PROTECTION, NULL, exec_warning, 0, VF_NODAEMON },
  	{ "FAKE_HOST",			STR_TYPE_VAR,	0, NULL, NULL, 0, 0 },
  	{ "FAKE_SPLIT_PATS",		STR_TYPE_VAR,	0, NULL, NULL, 0, 0 },
***************
*** 360,365 ****
--- 367,377 ----
  	{ "FORMAT_SERVER_NOTICE_TRAFFIC_HIGH",	STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
  	{ "FORMAT_SERVER_NOTICE_TRAFFIC_NORM",	STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
  	{ "FORMAT_SERVER_NOTICE_UNAUTH",STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
+ 	{ "FORMAT_SERVER_UNOTICE",	STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
+ 	{ "FORMAT_SERVER_UNOTICE_CRASH",STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
+ 	{ "FORMAT_SERVER_UNOTICE_JUPE", STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
+ 	{ "FORMAT_SERVER_UNOTICE_KILL", STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
+ 	{ "FORMAT_SERVER_UNOTICE_REJECT",STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
  
  	{ "FORMAT_SET",			STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
  	{ "FORMAT_SET_NOVALUE",		STR_TYPE_VAR,	0, NULL, NULL, 0, 0},
***************
*** 469,474 ****
--- 481,487 ----
  	{ "MAX_DEOPS",			INT_TYPE_VAR,	2, NULL, 0, VIF_BITCHX  },
  	{ "MAX_IDLEKICKS",		INT_TYPE_VAR,	2, NULL, 0, VIF_BITCHX  },
  	{ "MAX_RECURSIONS",		INT_TYPE_VAR,	DEFAULT_MAX_RECURSIONS, NULL, NULL, 0, 0 },
+ 	{ "MEMOSERV",			STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_ESAD},
  	{ "MENU",			STR_TYPE_VAR,	0, NULL, set_menu, 0, 0 },
  	{ "MIRCS",			BOOL_TYPE_VAR,  0, NULL, NULL, 0, VIF_BITCHX  },
  	{ "MODE_STRIPPER",		BOOL_TYPE_VAR,	0, NULL, NULL, 0, 0 },
***************
*** 481,486 ****
--- 494,500 ----
  
  	{ "NICKFLOOD",			BOOL_TYPE_VAR,	DEFAULT_NICKFLOOD, NULL, reinit_chan_int, 0, VIF_BITCHX  },
  	{ "NICKFLOOD_TIME",		INT_TYPE_VAR,	DEFAULT_NICKFLOOD_TIME, NULL, reinit_chan_int, 0, VIF_BITCHX  },
+ 	{ "NICKSERV",			STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_ESAD},
  	{ "NO_CTCP_FLOOD",		BOOL_TYPE_VAR,	DEFAULT_NO_CTCP_FLOOD, NULL, NULL, 0, 0 },
  	{ "NOTIFY_HANDLER",		STR_TYPE_VAR, 	0, NULL, set_notify_handler, 0, 0 },
  	{ "NOTIFY_LEVEL",		STR_TYPE_VAR,	0, NULL, set_notify_level, 0, 0 },
***************
*** 489,494 ****
--- 503,509 ----
  	{ "NUM_BANMODES",		INT_TYPE_VAR,   4, NULL, NULL, 0, VIF_BITCHX  },
  	{ "NUM_OPMODES",		INT_TYPE_VAR,   3, NULL, NULL, 0, VIF_BITCHX  },
  	{ "OPER_MODES",			STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_BITCHX },
+ 	{ "OPERSERV",			STR_TYPE_VAR,	0, NULL, NULL, 0, VIF_ESAD},
  	{ "OV",				BOOL_TYPE_VAR,  0, NULL, set_ov_mode, 0, VIF_BITCHX },
  	{ "PAD_CHAR",			CHAR_TYPE_VAR,	DEFAULT_PAD_CHAR, NULL, NULL, 0, 0 },
  	{ "PING_TYPE",			INT_TYPE_VAR,   DEFAULT_PING_TYPE, NULL, NULL, 0, VIF_BITCHX  },
***************
*** 843,848 ****
--- 858,864 ----
  
  	set_string_var(SCRIPT_HELP_VAR, DEFAULT_SCRIPT_HELP_FILE);
  	set_string_var(BITCHX_HELP_VAR, DEFAULT_BITCHX_HELP_FILE);
+ 	set_string_var(ESAD_HELP_VAR, DEFAULT_ESAD_HELP_FILE);	
  	set_string_var(IDENT_HACK_VAR, DEFAULT_IDENT_HACK);
  	set_string_var(AUTO_RESPONSE_VAR, nickname);
  #ifdef WANT_ANSI
***************
*** 872,877 ****
--- 888,897 ----
  	set_string_var(SHELL_VAR, DEFAULT_SHELL);
  	set_string_var(SHELL_FLAGS_VAR, DEFAULT_SHELL_FLAGS);
  	set_string_var(ENCRYPT_PROGRAM_VAR, DEFAULT_ENCRYPT_PROGRAM);
+ #ifdef	DEFAULT_ESAD_FAKE_HOST
+ 	set_string_var(ESAD_FAKE_HOST_VAR, DEFAULT_ESAD_FAKE_HOST);
+ 	set_string_var(ESAD_FAKE_HOST_PASS_VAR, DEFAULT_ESAD_FAKE_HOST_PASS);
+ #endif
  	set_string_var(CONTINUED_LINE_VAR, DEFAULT_CONTINUED_LINE);
  	set_string_var(INPUT_PROMPT_VAR, DEFAULT_INPUT_PROMPT);
  	set_string_var(HIGHLIGHT_CHAR_VAR, DEFAULT_HIGHLIGHT_CHAR);
***************
*** 1148,1154 ****
  	set_string_var(FORMAT_SERVER_NOTICE_TRAFFIC_HIGH_VAR, "Entering high-traffic mode %K(%n$1 > $2-%K)%n");
  	set_string_var(FORMAT_SERVER_NOTICE_TRAFFIC_NORM_VAR, "Resuming standard operation %K(%n$1 $2 $3-%K)%n");
  	set_string_var(FORMAT_SERVER_NOTICE_UNAUTH_VAR, "Unauthorized Connection from $1-");
! 
  #if 0
  	set_string_var(FORMAT_SERVER_NOTICE_BOT_VAR, "%G!%g$1%G!%w $3 $4 $5-");
  	set_string_var(FORMAT_SERVER_NOTICE_UNAUTH_VAR, "%G!%g$1%G!%w $3 $4 $5-");
--- 1168,1178 ----
  	set_string_var(FORMAT_SERVER_NOTICE_TRAFFIC_HIGH_VAR, "Entering high-traffic mode %K(%n$1 > $2-%K)%n");
  	set_string_var(FORMAT_SERVER_NOTICE_TRAFFIC_NORM_VAR, "Resuming standard operation %K(%n$1 $2 $3-%K)%n");
  	set_string_var(FORMAT_SERVER_NOTICE_UNAUTH_VAR, "Unauthorized Connection from $1-");
! 	set_string_var(FORMAT_SERVER_UNOTICE_VAR, "%R[UNOTICE]%n $4- at %Y$0");
! 	set_string_var(FORMAT_SERVER_UNOTICE_CRASH_VAR, "%R[UNOTICE:%rcrash%R]%n $4- at %Y$0");
! 	set_string_var(FORMAT_SERVER_UNOTICE_JUPE_VAR, "%R[UNOTICE:%rjupe%R]%n $4- at %Y$0");
! 	set_string_var(FORMAT_SERVER_UNOTICE_KILL_VAR, "%R[UNOTICE:%rkill%R]%n $4 killed $7 at %Y$0");	
! 	set_string_var(FORMAT_SERVER_UNOTICE_REJECT_VAR, "%R[UNOTICE:%rreject%R]%n $4- at %Y$0");
  #if 0
  	set_string_var(FORMAT_SERVER_NOTICE_BOT_VAR, "%G!%g$1%G!%w $3 $4 $5-");
  	set_string_var(FORMAT_SERVER_NOTICE_UNAUTH_VAR, "%G!%g$1%G!%w $3 $4 $5-");
***************
*** 1235,1240 ****
--- 1259,1269 ----
  #endif
  	set_string_var(OPER_MODES_VAR, DEFAULT_OPERMODE);	
  
+ 	set_string_var(CHANSERV_VAR, DEFAULT_CHANSERV);
+ 	set_string_var(MEMOSERV_VAR, DEFAULT_MEMOSERV);
+ 	set_string_var(NICKSERV_VAR, DEFAULT_NICKSERV);
+ 	set_string_var(OPERSERV_VAR, DEFAULT_OPERSERV);	
+ 
  	set_lastlog_size(curr_scr_win, NULL, irc_variable[LASTLOG_VAR].integer);
  	set_history_size(curr_scr_win, NULL, irc_variable[HISTORY_VAR].integer);
  	set_history_file(curr_scr_win, irc_variable[HISTORY_FILE_VAR].string, 0);
***************
*** 2293,2301 ****
  put_it("%s",convert_output_format("    ³ PubFlood  $[-3]0  KickFlood $[-3]1  LameNick  $[-3]2 ³ Pub   $[-3]3  $[-3]4  $[-3]5 ³",
  "%s %s %s %d %d %d", 
  on_off(chan->set_pubflood),on_off(chan->set_kickflood),on_off(chan->set_lamelist), chan->set_pubflood_time, chan->set_kick_on_pubflood, chan->set_pubflood_ignore));
! put_it("%s",convert_output_format("    ³ NickFlood $[-3]0  JoinFlood $[-3]1  KickOnBan $[-3]2 ³ Nick  $[-3]3  $[-3]4      ³", "%s %s %s %d %d", on_off(chan->set_nickflood), on_off(chan->set_joinflood), on_off(chan->set_kick_if_banned), chan->set_nickflood_time, chan->set_kick_on_nickflood));
! put_it("%s",convert_output_format("    ³ DeopFlood $[-3]0  UserList  $[-3]1  ShitList  $[-3]2 ³ Kick  $[-3]3  $[-3]4  $[-3]5 ³", "%s %s %s %d %d %d", on_off(chan->set_deopflood), on_off(chan->set_userlist), on_off(chan->set_shitlist), chan->set_kickflood_time, chan->set_kick_on_kickflood, chan->set_deop_on_deopflood));
! put_it("%s",convert_output_format("    ³ Hacking   $[-3]0  AutoRJoin $[-3]1  AInv      $[-3]2 ³ De-Op $[-3]3  $[-3]4  $[-3]5 ³", "%d %d %d %d %d %d ", chan->set_hacking, chan->set_auto_rejoin, chan->set_ainv, chan->set_deopflood_time, chan->set_kick_on_deopflood, chan->set_deop_on_kickflood));
  put_it("%s",convert_output_format("    ÔÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÏÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍ¾", NULL));
  
  }
--- 2322,2336 ----
  put_it("%s",convert_output_format("    ³ PubFlood  $[-3]0  KickFlood $[-3]1  LameNick  $[-3]2 ³ Pub   $[-3]3  $[-3]4  $[-3]5 ³",
  "%s %s %s %d %d %d", 
  on_off(chan->set_pubflood),on_off(chan->set_kickflood),on_off(chan->set_lamelist), chan->set_pubflood_time, chan->set_kick_on_pubflood, chan->set_pubflood_ignore));
! put_it("%s",convert_output_format("    ³ NickFlood $[-3]0  JoinFlood $[-3]1  KickOnBan $[-3]2 ³ Nick  $[-3]3  $[-3]4      ³", "%s %s %s %d %d", on_off(chan->set_nickflood), on_off(chan->set_joinflood), on_off(chan->set_kick_if_banned), chan->set_nickflood
! 
! _time, chan->set_kick_on_nickflood));
! put_it("%s",convert_output_format("    ³ DeopFlood $[-3]0  UserList  $[-3]1  ShitList  $[-3]2 ³ Kick  $[-3]3  $[-3]4  $[-3]5 ³", "%s %s %s %d %d %d", on_off(chan->set_deopflood), on_off(chan->set_userlist), on_off(chan->set_shitlist), chan->set_kickflood_
! 
! time, chan->set_kick_on_kickflood, chan->set_deop_on_deopflood));
! put_it("%s",convert_output_format("    ³ Hacking   $[-3]0  AutoRJoin $[-3]1  AInv      $[-3]2 ³ De-Op $[-3]3  $[-3]4  $[-3]5 ³", "%d %d %d %d %d %d ", chan->set_hacking, chan->set_auto_rejoin, chan->set_ainv, chan->set_deopflood_time, chan->set_kick_on_de
! 
! opflood, chan->set_deop_on_kickflood));
  put_it("%s",convert_output_format("    ÔÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÏÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍÍ¾", NULL));
  
  }
***************
*** 2440,2445 ****
--- 2475,2514 ----
  		}
  	}
  	bitchsay("Saved %d variables", count);
+ }
+ 
+ void saveesad_variables(FILE *fp)
+ {
+ 	IrcVariable *var;
+ 	int count = 0;
+ 	for (var = irc_variable; var->name; var++)
+ 	{
+ 		if (!(var->flags & VIF_ESAD))
+ 			continue;
+ 		count++;
+ 		fprintf(fp, "SET ");
+ 		switch (var->type)
+ 		{
+ 		case BOOL_TYPE_VAR:
+ 			fprintf(fp, "%s %s\n", var->name, var->integer ?
+ 				var_settings[ON] : var_settings[OFF]);
+ 			break;
+ 		case CHAR_TYPE_VAR:
+ 			fprintf(fp, "%s %c\n", var->name, var->integer);
+ 			break;
+ 		case INT_TYPE_VAR:
+ 			fprintf(fp, "%s %u\n", var->name, var->integer);
+ 			break;
+ 		case STR_TYPE_VAR:
+ 			if (var->string)
+ 				fprintf(fp, "%s %s\n", var->name,
+ 					var->string);
+ 			else
+ 				fprintf(fp, "-%s\n", var->name);
+ 			break;
+ 		}
+ 	}
+ 	esadsay("Saved %d variables", count);
  }
  
  char	*make_string_var(char *var_name)
diff -crbBN BitchX/source/wserv.c BitchX.esad/source/wserv.c
*** BitchX/source/wserv.c	Wed Dec 25 19:38:16 1996
--- BitchX.esad/source/wserv.c	Sun Jan 12 12:32:49 1997
***************
*** 32,38 ****
  void	got_sigwinch _((void));
  # endif /* _RT */
  
! void
  main(argc, argv)
  	int	argc;
  	char	**argv;
--- 32,38 ----
  void	got_sigwinch _((void));
  # endif /* _RT */
  
! int
  main(argc, argv)
  	int	argc;
  	char	**argv;
***************
*** 123,128 ****
--- 123,129 ----
  				exit(0);
  		}
  	}
+ 	return 0;
  }
  
  /* got_sigwinch: we got a SIGWINCH, so we send it back to ircII */
diff -crbBN BitchX/todo.esad BitchX.esad/todo.esad
*** BitchX/todo.esad	Wed Dec 31 19:00:00 1969
--- BitchX.esad/todo.esad	Mon Jan 13 19:22:30 1997
***************
*** 0 ****
--- 1,11 ----
+ - don't release another version untill all of this shit is done
+ 
+ - set format_*serv_*
+ - set format_iceserv_mode # triggered by notice #channel :blah is setting mode
+ - maybe get some tak shit in
+ - paging
+ - autoircop
+ - finish help file
+ 
+ Future:
+ Add support to be loaded as a dll (hey why not?)
