diff -cr lapack-old/BLAS/SRC/Makefile lapack/BLAS/SRC/Makefile
*** lapack-old/BLAS/SRC/Makefile	Fri Sep 23 11:10:44 1994
--- lapack/BLAS/SRC/Makefile	Sun Jan 28 22:20:37 1996
***************
*** 1,5 ****
--- 1,13 ----
  include ../../make.inc
  
+ # Uncomment these three lines for STATIC blas library
+ OPTS = -O2
+ NOOPT =
+ BLASTYPE = all-static
+ 
+ # Uncomment this line for DYNAMIC blas library
+ # BLASTYPE = all-shared
+ 
  #######################################################################
  #  This is the makefile to create a library for the BLAS.
  #  The files are grouped as follows:
***************
*** 50,57 ****
  #
  #######################################################################
  
! all: single double complex complex16
!  
  #---------------------------------------------------------
  #  Comment out the next 6 definitions if you already have
  #  the Level 1 BLAS.
--- 58,85 ----
  #
  #######################################################################
  
! all: $(BLASTYPE)
! 
! all-static: single double complex complex16
! 	rm -f ../../libblas.a
! 	ar rcv ../../libblas.a\
! 	$(ALLBLAS)\
! 	$(SBLAS1) $(SBLAS2) $(SBLAS3) \
! 	$(CBLAS1) $(CBLAS2) $(CBLAS3) \
! 	$(DBLAS1) $(DBLAS2) $(DBLAS3) \
! 	$(ZBLAS1) $(ZBLAS2) $(ZBLAS3) ../../INSTALL/etime.o
! 	ranlib ../../libblas.a
! 
! all-shared: single double complex complex16
! 	gcc -shared -Wl,-soname,libblas.so.2 -o ../../libblas.so.2.0.1\
! 	$(ALLBLAS)\
! 	$(SBLAS1) $(SBLAS2) $(SBLAS3) \
! 	$(CBLAS1) $(CBLAS2) $(CBLAS3) \
! 	$(DBLAS1) $(DBLAS2) $(DBLAS3) \
! 	$(ZBLAS1) $(ZBLAS2) $(ZBLAS3) ../../INSTALL/etime.o
! 	ln -sf ../../libblas.so.2.0.1 ../../libblas.so.2.0
! 	ln -sf ../../libblas.so.2.0 ../../libblas.so.2
! 
  #---------------------------------------------------------
  #  Comment out the next 6 definitions if you already have
  #  the Level 1 BLAS.
***************
*** 130,153 ****
  
  
  single: $(SBLAS1) $(ALLBLAS) $(SBLAS2) $(SBLAS3)
- 	$(ARCH) $(ARCHFLAGS) $(BLASLIB) $(SBLAS1) $(ALLBLAS) \
- 	$(SBLAS2) $(SBLAS3)
- 	$(RANLIB) $(BLASLIB)
  
  double: $(DBLAS1) $(ALLBLAS) $(DBLAS2) $(DBLAS3)
- 	$(ARCH) $(ARCHFLAGS) $(BLASLIB) $(DBLAS1) $(ALLBLAS) \
- 	$(DBLAS2) $(DBLAS3)
- 	$(RANLIB) $(BLASLIB)
  
  complex: $(CBLAS1) $(CB1AUX) $(ALLBLAS) $(CBLAS2) $(CBLAS3)
- 	$(ARCH) $(ARCHFLAGS) $(BLASLIB) $(CBLAS1) $(CB1AUX) \
- 	$(ALLBLAS) $(CBLAS2) $(CBLAS3)
- 	$(RANLIB) $(BLASLIB)
  
  complex16: $(ZBLAS1) $(ZB1AUX) $(ALLBLAS) $(ZBLAS2) $(ZBLAS3)
- 	$(ARCH) $(ARCHFLAGS) $(BLASLIB) $(ZBLAS1) $(ZB1AUX) \
- 	$(ALLBLAS) $(ZBLAS2) $(ZBLAS3)
- 	$(RANLIB) $(BLASLIB)
  
  FRC:
  	@FRC=$(FRC)
--- 158,169 ----
diff -cr lapack-old/INSTALL/Makefile lapack/INSTALL/Makefile
*** lapack-old/INSTALL/Makefile	Fri Sep 23 11:10:06 1994
--- lapack/INSTALL/Makefile	Sat Jan 20 22:19:01 1996
***************
*** 2,23 ****
  
  all:  testlsame testslamch testdlamch testsecond testdsecnd 
  
! testlsame:  lsame.o lsametst.o
! 	$(LOADER) -o testlsame lsame.o lsametst.o
  
! testslamch: slamch.o lsame.o slamchtst.o
! 	$(LOADER) -o testslamch slamch.o lsame.o slamchtst.o
  
! testdlamch: dlamch.o lsame.o dlamchtst.o
! 	$(LOADER) -o testdlamch dlamch.o lsame.o dlamchtst.o
  
! testsecond: second.o secondtst.o
! 	$(LOADER) -o testsecond second.o secondtst.o
  
! testdsecnd: dsecnd.o dsecndtst.o
! 	$(LOADER) -o testdsecnd dsecnd.o dsecndtst.o
  
  slamch.o: slamch.f ; $(FORTRAN) $(NOOPT) -c $<
  dlamch.o: dlamch.f ; $(FORTRAN) $(NOOPT) -c $<
  
  .f.o: ; $(FORTRAN) $(OPTS) -c $<
--- 2,24 ----
  
  all:  testlsame testslamch testdlamch testsecond testdsecnd 
  
! testlsame:  lsame.o lsametst.o etime.o
! 	$(LOADER) -o testlsame lsame.o lsametst.o etime.o 
  
! testslamch: slamch.o lsame.o slamchtst.o etime.o 
! 	$(LOADER) -o testslamch slamch.o lsame.o slamchtst.o etime.o
  
! testdlamch: dlamch.o lsame.o dlamchtst.o etime.o
! 	$(LOADER) -o testdlamch dlamch.o lsame.o dlamchtst.o etime.o
  
! testsecond: second.o secondtst.o etime.o
! 	$(LOADER) -o testsecond second.o secondtst.o etime.o
  
! testdsecnd: dsecnd.o dsecndtst.o etime.o
! 	$(LOADER) -o testdsecnd dsecnd.o dsecndtst.o etime.o
  
  slamch.o: slamch.f ; $(FORTRAN) $(NOOPT) -c $<
  dlamch.o: dlamch.f ; $(FORTRAN) $(NOOPT) -c $<
+ etime.o: etime.c
  
  .f.o: ; $(FORTRAN) $(OPTS) -c $<
diff -cr lapack-old/SRC/Makefile lapack/SRC/Makefile
*** lapack-old/SRC/Makefile	Fri Sep 23 11:11:49 1994
--- lapack/SRC/Makefile	Thu Jan 25 22:38:12 1996
***************
*** 248,273 ****
     zupmtr.o izmax1.o dzsum1.o
  
  all: single complex double complex16
  
! single: $(SLASRC) $(ALLAUX) $(SCLAUX) 
! 	$(ARCH) $(ARCHFLAGS) ../$(LAPACKLIB) $(SLASRC) $(ALLAUX) \
! 	$(SCLAUX)
! 	$(RANLIB) ../$(LAPACKLIB)
  
  complex: $(CLASRC) $(ALLAUX) $(SCLAUX)
- 	$(ARCH) $(ARCHFLAGS) ../$(LAPACKLIB) $(CLASRC) $(ALLAUX) \
- 	$(SCLAUX)
- 	$(RANLIB) ../$(LAPACKLIB)
  
  double: $(DLASRC) $(ALLAUX) $(DZLAUX)
- 	$(ARCH) $(ARCHFLAGS) ../$(LAPACKLIB) $(DLASRC) $(ALLAUX) \
- 	$(DZLAUX)
- 	$(RANLIB) ../$(LAPACKLIB)
  
  complex16: $(ZLASRC) $(ALLAUX) $(DZLAUX)
- 	$(ARCH) $(ARCHFLAGS) ../$(LAPACKLIB) $(ZLASRC) $(ALLAUX) \
- 	$(DZLAUX)
- 	$(RANLIB) ../$(LAPACKLIB)
  
  $(ALLAUX): $(FRC)
  $(SCLAUX): $(FRC)
--- 248,266 ----
     zupmtr.o izmax1.o dzsum1.o
  
  all: single complex double complex16
+ 	gcc -shared -Wl,-soname,liblapack.so.2 -o ../liblapack.so.2.0.1  \
+ 	$(ALLAUX)\
+ 	$(SLASRC) $(SCLAUX) $(CLASRC) $(DLASRC) $(DZLAUX) $(ZLASRC)
+ 	ln -sf ../liblapack.so.2.0.1 ../liblapack.so.2.0
+ 	ln -sf ../liblapack.so.2.0 ../liblapack.so.2
  
! single: $(SLASRC) $(ALLAUX) $(SCLAUX)
  
  complex: $(CLASRC) $(ALLAUX) $(SCLAUX)
  
  double: $(DLASRC) $(ALLAUX) $(DZLAUX)
  
  complex16: $(ZLASRC) $(ALLAUX) $(DZLAUX)
  
  $(ALLAUX): $(FRC)
  $(SCLAUX): $(FRC)
diff -cr lapack-old/TESTING/EIG/Makefile lapack/TESTING/EIG/Makefile
*** lapack-old/TESTING/EIG/Makefile	Fri Sep 23 11:31:46 1994
--- lapack/TESTING/EIG/Makefile	Sat Jan 20 22:19:01 1996
***************
*** 49,55 ****
     ilaenv.o \
     xerbla.o \
     xlaenv.o \
!    chkxer.o
  
  SCIGTST = slafts.o slahd2.o slasum.o slatb9.o sstech.o sstect.o \
     ssvdch.o ssvdct.o ssxt1.o
--- 49,55 ----
     ilaenv.o \
     xerbla.o \
     xlaenv.o \
!    chkxer.o 
  
  SCIGTST = slafts.o slahd2.o slasum.o slatb9.o sstech.o sstect.o \
     ssvdch.o ssvdct.o ssxt1.o
***************
*** 128,150 ****
  
  ../xeigtsts: $(SEIGTST) $(SCIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(SEIGTST) $(SCIGTST) $(AEIGTST) ../../$(TMGLIB) \
! 	  ../../$(LAPACKLIB) $(BLASLIB)
  
  ../xeigtstc: $(CEIGTST) $(SCIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(CEIGTST) $(SCIGTST) $(AEIGTST) ../../$(TMGLIB) \
! 	  ../../$(LAPACKLIB) $(BLASLIB)
  
  ../xeigtstd: $(DEIGTST) $(DZIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(DEIGTST) $(DZIGTST) $(AEIGTST) ../../$(TMGLIB) \
! 	  ../../$(LAPACKLIB) $(BLASLIB)
  
  ../xeigtstz: $(ZEIGTST) $(DZIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(ZEIGTST) $(DZIGTST) $(AEIGTST) ../../$(TMGLIB) \
! 	  ../../$(LAPACKLIB) $(BLASLIB)
  
  $(AEIGTST): $(FRC)
  $(SCIGTST): $(FRC)
--- 128,150 ----
  
  ../xeigtsts: $(SEIGTST) $(SCIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(SEIGTST) $(SCIGTST) $(AEIGTST) $(TMGLIB) \
! 	  $(LAPACKLIB) $(BLASLIB)
  
  ../xeigtstc: $(CEIGTST) $(SCIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(CEIGTST) $(SCIGTST) $(AEIGTST) $(TMGLIB) \
! 	  $(LAPACKLIB) $(BLASLIB)
  
  ../xeigtstd: $(DEIGTST) $(DZIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(DEIGTST) $(DZIGTST) $(AEIGTST) $(TMGLIB) \
! 	  $(LAPACKLIB) $(BLASLIB)
  
  ../xeigtstz: $(ZEIGTST) $(DZIGTST) $(AEIGTST) ; \
            $(LOADER) $(LOADOPTS) -o $@ \
!           $(ZEIGTST) $(DZIGTST) $(AEIGTST) $(TMGLIB) \
! 	  $(LAPACKLIB) $(BLASLIB)
  
  $(AEIGTST): $(FRC)
  $(SCIGTST): $(FRC)
diff -cr lapack-old/TESTING/LIN/Makefile lapack/TESTING/LIN/Makefile
*** lapack-old/TESTING/LIN/Makefile	Fri Sep 23 11:26:31 1994
--- lapack/TESTING/LIN/Makefile	Sat Jan 20 22:19:02 1996
***************
*** 42,48 ****
  
  ALINTST = \
     aladhd.o alaerh.o alaesm.o alahd.o alareq.o \
!    alasum.o alasvm.o chkxer.o ilaenv.o xlaenv.o xerbla.o
  
  SCLNTST= slaord.o
  
--- 42,48 ----
  
  ALINTST = \
     aladhd.o alaerh.o alaesm.o alahd.o alareq.o \
!    alasum.o alasvm.o chkxer.o ilaenv.o xlaenv.o xerbla.o 
  
  SCLNTST= slaord.o
  
***************
*** 177,195 ****
  
  ../xlintsts : $(ALINTST) $(SLINTST) $(SCLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(SCLNTST) $(SLINTST) \
!         ../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB)  -o $@
  
  ../xlintstc : $(ALINTST) $(CLINTST) $(SCLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(SCLNTST) $(CLINTST) \
!         ../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB)  -o $@
   
  ../xlintstd : $(ALINTST) $(DLINTST) $(DZLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(DZLNTST) $(DLINTST) \
!         ../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB)  -o $@
   
  ../xlintstz : $(ALINTST) $(ZLINTST) $(DZLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(DZLNTST) $(ZLINTST) \
!         ../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB)  -o $@
   
  $(ALINTST): $(FRC)
  $(SCLNTST): $(FRC)
--- 177,195 ----
  
  ../xlintsts : $(ALINTST) $(SLINTST) $(SCLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(SCLNTST) $(SLINTST) \
!         $(TMGLIB) $(LAPACKLIB) $(BLASLIB)  -o $@
  
  ../xlintstc : $(ALINTST) $(CLINTST) $(SCLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(SCLNTST) $(CLINTST) \
!         $(TMGLIB) $(LAPACKLIB) $(BLASLIB)  -o $@
   
  ../xlintstd : $(ALINTST) $(DLINTST) $(DZLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(DZLNTST) $(DLINTST) \
!         $(TMGLIB) $(LAPACKLIB) $(BLASLIB)  -o $@
   
  ../xlintstz : $(ALINTST) $(ZLINTST) $(DZLNTST)
  	$(LOADER) $(LOADOPTS)  $(ALINTST) $(DZLNTST) $(ZLINTST) \
!         $(TMGLIB) $(LAPACKLIB) $(BLASLIB)  -o $@
   
  $(ALINTST): $(FRC)
  $(SCLNTST): $(FRC)
diff -cr lapack-old/TESTING/MATGEN/Makefile lapack/TESTING/MATGEN/Makefile
*** lapack-old/TESTING/MATGEN/Makefile	Fri Sep 23 11:22:58 1994
--- lapack/TESTING/MATGEN/Makefile	Sat Jan 20 22:19:02 1996
***************
*** 57,78 ****
     zlarnd.o zlatm2.o zlatm3.o
  
  all: single complex double complex16
  
  single: $(SMATGEN) $(SCATGEN)
- 	$(ARCH) $(ARCHFLAGS) ../../$(TMGLIB) $(SMATGEN) $(SCATGEN)
- 	$(RANLIB) ../../$(TMGLIB)
  
  complex: $(CMATGEN) $(SCATGEN)
- 	$(ARCH) $(ARCHFLAGS) ../../$(TMGLIB) $(CMATGEN) $(SCATGEN)
- 	$(RANLIB) ../../$(TMGLIB)
  
  double: $(DMATGEN) $(DZATGEN)
- 	$(ARCH) $(ARCHFLAGS) ../../$(TMGLIB) $(DMATGEN) $(DZATGEN)
- 	$(RANLIB) ../../$(TMGLIB)
  
  complex16: $(ZMATGEN) $(DZATGEN)
- 	$(ARCH) $(ARCHFLAGS) ../../$(TMGLIB) $(ZMATGEN) $(DZATGEN)
- 	$(RANLIB) ../../$(TMGLIB)
  
  $(SCATGEN): $(FRC)
  $(SMATGEN): $(FRC)
--- 57,74 ----
     zlarnd.o zlatm2.o zlatm3.o
  
  all: single complex double complex16
+ 	gcc -shared -Wl,-soname,libtmglib.so.2 -o ../../libtmglib.so.2.0  \
+ 	$(SMATGEN) $(SCATGEN) $(CMATGEN) $(DMATGEN) $(DZATGEN) $(ZMATGEN)
+ 	ln -sf ../../libtmglib.so.2.0 ../../libtmglib.so.2
+ 	ln -sf ../../libtmglib.so.2 ../../libtmglib.so
  
  single: $(SMATGEN) $(SCATGEN)
  
  complex: $(CMATGEN) $(SCATGEN)
  
  double: $(DMATGEN) $(DZATGEN)
  
  complex16: $(ZMATGEN) $(DZATGEN)
  
  $(SCATGEN): $(FRC)
  $(SMATGEN): $(FRC)
diff -cr lapack-old/TIMING/EIG/EIGSRC/Makefile lapack/TIMING/EIG/EIGSRC/Makefile
*** lapack-old/TIMING/EIG/EIGSRC/Makefile	Fri Sep 23 11:37:00 1994
--- lapack/TIMING/EIG/EIGSRC/Makefile	Sat Jan 20 22:19:02 1996
***************
*** 73,94 ****
     zpteqr.o zstedc.o zstein.o zsteqr.o ztgevc.o ztrevc.o
  
  all: single complex double complex16
  
  single: $(SEIGSRC) $(SCIGSRC)
- 	$(ARCH) $(ARCHFLAGS) ../$(EIGSRCLIB) $(SEIGSRC) $(SCIGSRC)
- 	$(RANLIB) ../$(EIGSRCLIB)
  
  complex: $(CEIGSRC) $(SCIGSRC)
- 	$(ARCH) $(ARCHFLAGS) ../$(EIGSRCLIB) $(CEIGSRC) $(SCIGSRC)
- 	$(RANLIB) ../$(EIGSRCLIB)
  
  double: $(DEIGSRC) $(DZIGSRC)
- 	$(ARCH) $(ARCHFLAGS) ../$(EIGSRCLIB) $(DEIGSRC) $(DZIGSRC)
- 	$(RANLIB) ../$(EIGSRCLIB)
  
  complex16: $(ZEIGSRC) $(DZIGSRC)
- 	$(ARCH) $(ARCHFLAGS) ../$(EIGSRCLIB) $(ZEIGSRC) $(DZIGSRC)
- 	$(RANLIB) ../$(EIGSRCLIB)
  
  $(SCIGSRC): $(FRC)
  $(DZIGSRC): $(FRC)
--- 73,91 ----
     zpteqr.o zstedc.o zstein.o zsteqr.o ztgevc.o ztrevc.o
  
  all: single complex double complex16
+ 	gcc -shared -Wl,-soname,libeigsrc.so.2.0 -o ../../../libeigsrc.so.2.0 \
+ 	$(SEIGSRC) $(SCIGSRC)\
+         $(CEIGSRC) $(DEIGSRC) $(DZIGSRC) $(ZEIGSRC)
+ 	ln -sf ../../../libeigsrc.so.2.0 ../../../libeigsrc.so.2
+ 	ln -sf ../../../libeigsrc.so.2 ../../../libeigsrc.so
  
  single: $(SEIGSRC) $(SCIGSRC)
  
  complex: $(CEIGSRC) $(SCIGSRC)
  
  double: $(DEIGSRC) $(DZIGSRC)
  
  complex16: $(ZEIGSRC) $(DZIGSRC)
  
  $(SCIGSRC): $(FRC)
  $(DZIGSRC): $(FRC)
diff -cr lapack-old/TIMING/EIG/Makefile lapack/TIMING/EIG/Makefile
*** lapack-old/TIMING/EIG/Makefile	Fri Sep 23 11:35:59 1994
--- lapack/TIMING/EIG/Makefile	Sat Jan 20 22:19:02 1996
***************
*** 41,47 ****
  #
  #######################################################################
  
! AEIGTIM = atimin.o ilaenv.o xlaenv.o
  
  SCIGTIM = \
     smflop.o sopla.o sopla2.o \
--- 41,47 ----
  #
  #######################################################################
  
! AEIGTIM = atimin.o ilaenv.o xlaenv.o 
  
  SCIGTIM = \
     smflop.o sopla.o sopla2.o \
***************
*** 72,93 ****
  
  ../xeigtims:$(AEIGTIM) $(SCIGTIM) $(SEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(SCIGTIM) $(SEIGTIM) \
!         ../../$(TMGLIB) $(EIGSRCLIB) ../../$(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtims
  
  ../xeigtimc:$(AEIGTIM) $(SCIGTIM) $(CEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(SCIGTIM) $(CEIGTIM) \
!         ../../$(TMGLIB) $(EIGSRCLIB) ../../$(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtimc
  
  ../xeigtimd:$(AEIGTIM) $(DZIGTIM) $(DEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(DZIGTIM) $(DEIGTIM) \
!         ../../$(TMGLIB) $(EIGSRCLIB) ../../$(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtimd
  
  ../xeigtimz:$(AEIGTIM) $(DZIGTIM) $(ZEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(DZIGTIM) $(ZEIGTIM) \
!         ../../$(TMGLIB) $(EIGSRCLIB) ../../$(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtimz
  
  $(AEIGTIM): $(FRC)
--- 72,93 ----
  
  ../xeigtims:$(AEIGTIM) $(SCIGTIM) $(SEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(SCIGTIM) $(SEIGTIM) \
!         $(TMGLIB) $(EIGSRCLIB) $(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtims
  
  ../xeigtimc:$(AEIGTIM) $(SCIGTIM) $(CEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(SCIGTIM) $(CEIGTIM) \
!         $(TMGLIB) $(EIGSRCLIB) $(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtimc
  
  ../xeigtimd:$(AEIGTIM) $(DZIGTIM) $(DEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(DZIGTIM) $(DEIGTIM) \
!         $(TMGLIB) $(EIGSRCLIB) $(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtimd
  
  ../xeigtimz:$(AEIGTIM) $(DZIGTIM) $(ZEIGTIM)
  	$(LOADER) $(LOADOPTS)  $(AEIGTIM) $(DZIGTIM) $(ZEIGTIM) \
!         $(TMGLIB) $(EIGSRCLIB) $(LAPACKLIB) \
  	$(BLASLIB) -o ../xeigtimz
  
  $(AEIGTIM): $(FRC)
diff -cr lapack-old/TIMING/LIN/Makefile lapack/TIMING/LIN/Makefile
*** lapack-old/TIMING/LIN/Makefile	Fri Sep 23 11:34:45 1994
--- lapack/TIMING/LIN/Makefile	Sat Jan 20 22:19:02 1996
***************
*** 85,103 ****
  
  ../xlintims : $(ALINTIM) $(SCINTIM) $(SLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(SCINTIM) $(SLINTIM) \
! 	../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB)  -o $@
  
  ../xlintimc : $(ALINTIM) $(SCINTIM) $(CLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(SCINTIM) $(CLINTIM) \
! 	../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB) -o $@
  
  ../xlintimd : $(ALINTIM) $(DZINTIM) $(DLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(DZINTIM) $(DLINTIM) \
! 	../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB)  -o $@
  
  ../xlintimz : $(ALINTIM) $(DZINTIM) $(ZLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(DZINTIM) $(ZLINTIM) \
! 	../../$(TMGLIB) ../../$(LAPACKLIB) $(BLASLIB)  -o $@
  
  $(ALINTIM): $(FRC)
  $(SCINTIM): $(FRC)
--- 85,103 ----
  
  ../xlintims : $(ALINTIM) $(SCINTIM) $(SLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(SCINTIM) $(SLINTIM) \
! 	$(TMGLIB) $(LAPACKLIB) $(BLASLIB)  -o $@
  
  ../xlintimc : $(ALINTIM) $(SCINTIM) $(CLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(SCINTIM) $(CLINTIM) \
! 	$(TMGLIB) $(LAPACKLIB) $(BLASLIB) -o $@
  
  ../xlintimd : $(ALINTIM) $(DZINTIM) $(DLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(DZINTIM) $(DLINTIM) \
! 	$(TMGLIB) $(LAPACKLIB) $(BLASLIB)  -o $@
  
  ../xlintimz : $(ALINTIM) $(DZINTIM) $(ZLINTIM)
  	$(LOADER) $(LOADOPTS)  $(ALINTIM) $(DZINTIM) $(ZLINTIM) \
! 	$(TMGLIB) $(LAPACKLIB) $(BLASLIB)  -o $@
  
  $(ALINTIM): $(FRC)
  $(SCINTIM): $(FRC)
diff -cr lapack-old/TIMING/Makefile lapack/TIMING/Makefile
*** lapack-old/TIMING/Makefile	Fri Sep 23 11:34:00 1994
--- lapack/TIMING/Makefile	Sat Jan 20 22:19:02 1996
***************
*** 397,415 ****
  
  
  xeigtims:
! 	cd EIG/EIGSRC ; $(MAKE) single
  	cd EIG ; $(MAKE) single
  
  xeigtimc:
! 	cd EIG/EIGSRC ; $(MAKE) complex
  	cd EIG ; $(MAKE) complex
  
  xeigtimd:
! 	cd EIG/EIGSRC ; $(MAKE) double
  	cd EIG ; $(MAKE) double
  
  xeigtimz:
! 	cd EIG/EIGSRC ; $(MAKE) complex16
  	cd EIG ; $(MAKE) complex16
  
  clean:
--- 397,415 ----
  
  
  xeigtims:
! 	cd EIG/EIGSRC ; $(MAKE) all
  	cd EIG ; $(MAKE) single
  
  xeigtimc:
! 	cd EIG/EIGSRC ; $(MAKE) all
  	cd EIG ; $(MAKE) complex
  
  xeigtimd:
! 	cd EIG/EIGSRC ; $(MAKE) all
  	cd EIG ; $(MAKE) double
  
  xeigtimz:
! 	cd EIG/EIGSRC ; $(MAKE) all
  	cd EIG ; $(MAKE) complex16
  
  clean:
diff -cr lapack-old/make.inc lapack/make.inc
*** lapack-old/make.inc	Fri Sep 23 11:09:30 1994
--- lapack/make.inc	Sat Jan 20 22:19:02 1996
***************
*** 6,12 ****
  #
  #  The machine (platform) identifier to append to the library names
  #
! PLAT = _sun4
  #  
  #  Modify the FORTRAN and OPTS definitions to refer to the
  #  compiler and desired compiler options for your machine.  NOOPT
--- 6,12 ----
  #
  #  The machine (platform) identifier to append to the library names
  #
! PLAT = linux-elf
  #  
  #  Modify the FORTRAN and OPTS definitions to refer to the
  #  compiler and desired compiler options for your machine.  NOOPT
***************
*** 14,38 ****
  #  selected.  Define LOADER and LOADOPTS to refer to the loader and 
  #  desired load options for your machine.
  #
! FORTRAN  = f77
! OPTS     = -O -u
! NOOPT    = -u
! LOADER   = f77
! LOADOPTS =
  #
  #  The archiver and the flag(s) to use when building archive (library)
  #  If you system has no ranlib, set RANLIB = echo.
  #
! ARCH     = ar
! ARCHFLAGS= cr
  RANLIB   = ranlib
  #
  #  The location of the libraries to which you will link.  (The 
  #  machine-specific, optimized BLAS library should be used whenever
  #  possible.)
  #
! BLASLIB      = ../../blas$(PLAT).a
! #BLASLIB     = -lblas
! LAPACKLIB    = lapack$(PLAT).a
! TMGLIB       = tmglib$(PLAT).a
! EIGSRCLIB    = eigsrc$(PLAT).a
--- 14,41 ----
  #  selected.  Define LOADER and LOADOPTS to refer to the loader and 
  #  desired load options for your machine.
  #
! FORTRAN  = g77
! OPTS     = -O2 -u -fPIC
! NOOPT    = -u -fPIC
! # You may have to change the -L to wherever you build stuff
! LOADER   = g77 -lm -lieee -L/usr/local/src/LAPACK
! LOADOPTS = 
  #
  #  The archiver and the flag(s) to use when building archive (library)
  #  If you system has no ranlib, set RANLIB = echo.
  #
! ARCH     = gcc 
! ARCHFLAGS= -shared -Wl,-soname,libfoo.so.2 
  RANLIB   = ranlib
  #
  #  The location of the libraries to which you will link.  (The 
  #  machine-specific, optimized BLAS library should be used whenever
  #  possible.)
  #
! BLASLIB      = -lblas
! LAPACKLIB    = -llapack
! TMGLIB       = -ltmglib
! EIGSRCLIB    = -leigsrc
! 
! 
! 
*** /dev/null	Sun Jul 17 18:46:18 1994
--- lapack/INSTALL/etime.c	Sat Jan 20 22:22:49 1996
***************
*** 0 ****
--- 1,27 ----
+ /*
+   From David Klein (klein@sunrise.huji.ac.za)
+ */
+ 
+ #include <sys/times.h>
+ #include <unistd.h>
+ 
+ #include "f2c.h"
+ 
+ doublereal etime_(tarray)
+ real *tarray;
+ {
+   struct tms buf;
+   time_t ticks;
+ 
+     /* System generated locals */
+     real ret_val;
+ 
+     /* Parameter adjustments */
+     --tarray;
+ 
+     /* Function Body */
+   ticks=times(&buf);
+   tarray[1]=buf.tms_utime/100.0;
+   ret_val=tarray[1];
+   return ret_val;
+ }
