*** /usr/X11R6/lib/Server/drivers/vga256/et4000/et4_driver.c	Sun Jan 29 05:14:06 1995
--- /home/kmg/XFree86-3.1.1-1/xc/programs/Xserver/hw/xfree86/vga256/drivers/et4000/et4_driver.c	Sun Jun 18 16:12:31 1995
***************
*** 593,603 ****
  
    /*
     * we need this here , cause we MUST disable the ROM SYNC feature
     */
    outb(vgaIOBase + 4, 0x34); temp1 = inb(vgaIOBase + 5);
!   outb(vgaIOBase + 5, temp1 & 0x0F);
    temp2 = inb(0x3CD); outb(0x3CD, 0x00); /* segment select */
  
    save = (vgaET4000Ptr)vgaHWSave((vgaHWPtr)save, sizeof(vgaET4000Rec));
    save->Compatibility = temp1;
    save->SegSel = temp2;
--- 593,603 ----
  
    /*
     * we need this here , cause we MUST disable the ROM SYNC feature
     */
    outb(vgaIOBase + 4, 0x34); temp1 = inb(vgaIOBase + 5);
!   outb(vgaIOBase + 5, temp1 & 0x1F);
    temp2 = inb(0x3CD); outb(0x3CD, 0x00); /* segment select */
  
    save = (vgaET4000Ptr)vgaHWSave((vgaHWPtr)save, sizeof(vgaET4000Rec));
    save->Compatibility = temp1;
    save->SegSel = temp2;
