--- local_curses.h.orig	Tue Jun  4 23:38:30 1996
+++ local_curses.h	Tue Jun  4 23:39:25 1996
@@ -32,7 +32,6 @@
 
 #include <stdio.h>   
 #include <unctrl.h>
-#include <stdarg.h>
 #ifdef _XOPEN_SOURCE_EXTENDED
 #include <stddef.h>	/* we want wchar_t */
 #endif /* _XOPEN_SOURCE_EXTENDED */
@@ -621,7 +620,6 @@
 #ifdef _XOPEN_SOURCE_EXTENDED
 extern int slk_wset(int, wchar_t *, int);		/* missing */
 #endif /* _XOPEN_SOURCE_EXTENDED */
-extern int standout(void);				/* generated */
 extern int standend(void);				/* generated */
 extern int start_color(void);				/* implemented */
 extern WINDOW *subpad(WINDOW *, int, int, int, int);	/* implemented */
