*** drivers/sound/lowlevel/Makefile.old	Mon Sep  9 01:50:21 1996
--- drivers/sound/lowlevel/Makefile	Wed Oct 30 21:00:40 1996
***************
*** 6,11 ****
--- 6,14 ----
  ifdef CONFIG_ACI_MIXER
  OBJS := $(OBJS) aci.o
  endif
+ ifdef CONFIG_AWE32_SYNTH
+ OBJS := $(OBJS) awe_wave.o
+ endif
  endif
  
  lowlevel.o:	$(OBJS)
*** drivers/sound/lowlevel/init.c.old	Tue May 28 23:45:17 1996
--- drivers/sound/lowlevel/init.c	Wed Oct 30 21:02:02 1996
***************
*** 7,12 ****
--- 7,14 ----
  #ifdef CONFIG_LOWLEVEL_SOUND
  extern int attach_aci(void);
  extern void unload_aci(void);
+ extern int attach_awe(void);
+ extern void unload_awe(void);
  
  void
  sound_init_lowlevel_drivers(void)
***************
*** 14,19 ****
--- 16,24 ----
  #ifdef CONFIG_ACI_MIXER
     attach_aci();
  #endif
+ #ifdef CONFIG_AWE32_SYNTH
+    attach_awe();
+ #endif
  }
  
  void
***************
*** 21,26 ****
--- 26,34 ----
  {
  #ifdef CONFIG_ACI_MIXER
     unload_aci();
+ #endif
+ #ifdef CONFIG_AWE32_SYNTH
+    unload_awe();
  #endif
  }
  #endif
*** drivers/sound/lowlevel/Config.tmpl.old	Thu May 16 17:05:56 1996
--- drivers/sound/lowlevel/Config.tmpl	Wed Oct 30 21:02:02 1996
***************
*** 2,5 ****
--- 2,6 ----
  
  if [ "$CONFIG_LOWLEVEL_SOUND" = "y" ]; then
     bool 'ACI mixer (miroPCM12)' CONFIG_ACI_MIXER
+    bool 'AWE32 synth' CONFIG_AWE32_SYNTH
  fi
*** drivers/sound/Config.in.old	Wed Dec 11 18:19:41 1996
--- drivers/sound/Config.in	Thu Dec  5 15:56:44 1996
***************
*** 250,253 ****
--- 250,254 ----
  
  if [ "$CONFIG_LOWLEVEL_SOUND" = "y" ]; then
     bool 'ACI mixer (miroPCM12)' CONFIG_ACI_MIXER
+    bool 'AWE32 wave synth' CONFIG_AWE32_SYNTH
  fi
*** Documentation/Configure.help.old	Sat Oct  5 18:13:56 1996
--- Documentation/Configure.help	Wed Oct 30 21:02:03 1996
***************
*** 3685,3690 ****
--- 3685,3696 ----
    This is support for the soundcards on Sun workstations. The code
    does not exist yet, so you might as well say N here.
  
+ AWE32 support
+ CONFIG_AWE32_SYNTH
+   Answering Y enables Soundblaster AWE32 support. This includes
+   emu8k support as well as on-card memory control. If unsure,
+   answer N.
+ 
  Kernel profiling support
  CONFIG_PROFILE
    This is for kernel hackers who want to know how much time the kernel
