*** configure.c.old	Fri Aug 23 21:21:22 1996
--- configure.c	Thu Aug 22 23:12:18 1996
***************
*** 60,66 ****
  #define OPT_YM3812_AUTO	19
  #define OPT_YM3812	20
  #define OPT_SEQUENCER	21
! #define OPT_LAST	21	/* Last defined OPT number */
  
  #define ANY_DEVS (B(OPT_AUDIO)|B(OPT_MIDI)|B(OPT_SEQUENCER)|B(OPT_GUS)| \
  		  B(OPT_MPU401)|B(OPT_PSS)|B(OPT_GUS16)|B(OPT_GUSMAX)| \
--- 60,67 ----
  #define OPT_YM3812_AUTO	19
  #define OPT_YM3812	20
  #define OPT_SEQUENCER	21
! #define OPT_AWE32	22
! #define OPT_LAST	22	/* Last defined OPT number */
  
  #define ANY_DEVS (B(OPT_AUDIO)|B(OPT_MIDI)|B(OPT_SEQUENCER)|B(OPT_GUS)| \
  		  B(OPT_MPU401)|B(OPT_PSS)|B(OPT_GUS16)|B(OPT_GUSMAX)| \
***************
*** 132,138 ****
    {MIDI_CARDS, 0, "MIDI", 1, 0, 1},
    {B (OPT_ADLIB), 0, "YM3812_AUTO", 0, OPT_YM3812, 0},
    {B (OPT_PSS) | B (OPT_SB) | B (OPT_PAS) | B (OPT_ADLIB) | B (OPT_MSS) | B (OPT_PSS), B (OPT_YM3812_AUTO), "YM3812", 1, 0, 1},
!   {B (OPT_MIDI) | B (OPT_YM3812) | B (OPT_YM3812_AUTO) | B (OPT_GUS), 0, "SEQUENCER", 0, 0, 1}
  };
  
  char           *questions[] =
--- 133,140 ----
    {MIDI_CARDS, 0, "MIDI", 1, 0, 1},
    {B (OPT_ADLIB), 0, "YM3812_AUTO", 0, OPT_YM3812, 0},
    {B (OPT_PSS) | B (OPT_SB) | B (OPT_PAS) | B (OPT_ADLIB) | B (OPT_MSS) | B (OPT_PSS), B (OPT_YM3812_AUTO), "YM3812", 1, 0, 1},
!   {B (OPT_MIDI) | B (OPT_YM3812) | B (OPT_YM3812_AUTO) | B (OPT_GUS), 0, "SEQUENCER", 0, 0, 1},
!   {B (OPT_SB) | B (OPT_SBPRO), B (OPT_PAS), "AWE32", 1, 0, 0},
  };
  
  char           *questions[] =
***************
*** 160,165 ****
--- 162,168 ----
    "This should not be asked",
    "FM synthesizer (YM3812/OPL-3) support",
    "/dev/sequencer support",
+   "AWE32 Synth support",
    "Is the sky really falling"
  };
  
*** dev_table.h.old	Fri Aug 23 21:22:01 1996
--- dev_table.h	Wed Aug 14 02:23:34 1996
***************
*** 288,293 ****
--- 288,296 ----
  		{SNDCARD_SB16MIDI,"SB16 MIDI",	attach_sb16midi, probe_sb16midi},
  #endif
  #endif
+ #ifndef EXCLUDE_AWE32
+ 		{SNDCARD_AWE32,	"AWE32 Synth",	attach_awe_obsolete, probe_awe_obsolete},
+ #endif
  #ifndef EXCLUDE_GUS16
  		{SNDCARD_GUS16,	"Ultrasound 16-bit opt.",	attach_gus_db16, probe_gus_db16},
  #endif
***************
*** 384,389 ****
--- 387,395 ----
  #endif
  #ifndef EXCLUDE_MIDI
  		{SNDCARD_SB16MIDI,{SB16MIDI_BASE, SBC_IRQ, 0}, SND_DEFAULT_ENABLE},
+ #endif
+ #ifndef EXCLUDE_AWE32
+ 		{SNDCARD_AWE32, {SBC_BASE, 0, 0}, SND_DEFAULT_ENABLE},
  #endif
  #endif
  
*** makefile.linux.old	Fri Aug 23 21:22:37 1996
--- makefile.linux	Tue Aug 13 23:27:41 1996
***************
*** 22,28 ****
  	 gus_midi.o gus_vol.o patmgr.o sb_mixer.o sb16_dsp.o sb_midi.o \
  	 sb16_midi.o sound_switch.o midi_synth.o uart6850.o sound_timer.o \
  	 sys_timer.o ics2101.o ad1848.o pss.o sscape.o trix.o aedsp16.o \
! 	 mad16.o
  
  all:	/usr/include/linux/soundcard.h local.h sound.a
  
--- 22,28 ----
  	 gus_midi.o gus_vol.o patmgr.o sb_mixer.o sb16_dsp.o sb_midi.o \
  	 sb16_midi.o sound_switch.o midi_synth.o uart6850.o sound_timer.o \
  	 sys_timer.o ics2101.o ad1848.o pss.o sscape.o trix.o aedsp16.o \
! 	 mad16.o awe_wave.o
  
  all:	/usr/include/linux/soundcard.h local.h sound.a
  
*** sound_calls.h.old	Fri Aug 23 21:23:00 1996
--- sound_calls.h	Tue Aug 13 23:39:01 1996
***************
*** 273,275 ****
--- 273,279 ----
  int probe_mad16 (struct address_info *hw_config);
  long attach_mad16_mpu (long mem_start, struct address_info *hw_config);
  int probe_mad16_mpu (struct address_info *hw_config);
+ 
+ /* from awe_wave.c */
+ long attach_awe_obsolete(long mem_start, struct address_info *hw_config);
+ int probe_awe_obsolete(struct address_info *hw_config);
*** soundcard.h.old	Fri Aug 23 21:21:33 1996
--- soundcard.h	Tue Aug 13 23:57:44 1996
***************
*** 75,80 ****
--- 75,81 ----
  #define SNDCARD_TRXPRO_MPU	18
  #define SNDCARD_MAD16		19
  #define SNDCARD_MAD16_MPU	20
+ #define SNDCARD_AWE32		21
  
  /***********************************
   * IOCTL Commands for /dev/sequencer
***************
*** 506,511 ****
--- 507,513 ----
  #define FM_TYPE_OPL3			0x01
  
  #define SAMPLE_TYPE_GUS			0x10
+ #define SAMPLE_TYPE_AWE32		0x20
  
  		int	perc_mode;	/* No longer supported */
  		int	nr_voices;
