diff -c temp/scm/findexec.c temp/nscm/findexec.c
*** temp/scm/findexec.c	Sun Mar 17 23:16:26 1996
--- temp/nscm/findexec.c	Thu Mar 21 08:51:27 1996
***************
*** 37,46 ****
     filename.  A new copy of the complete path name of that file is
     returned.  This new string may be disposed by free() later on.  */
  
  #include <sys/file.h>
  #include <sys/param.h>
! #include <strings.h>
! #ifdef linux
  # include <stdlib.h>
  # include <sys/stat.h>
  # include <unistd.h>     /* for X_OK define */
--- 37,47 ----
     filename.  A new copy of the complete path name of that file is
     returned.  This new string may be disposed by free() later on.  */
  
+ #include "scm.h"
+ 
  #include <sys/file.h>
  #include <sys/param.h>
! #if defined(linux) || defined(__svr4__)
  # include <stdlib.h>
  # include <sys/stat.h>
  # include <unistd.h>     /* for X_OK define */
***************
*** 116,122 ****
      if (*p) p++;
  
      if (name[0] == '.' && name[1] == 0)
!       getwd(name);
  
      else if (name[0]=='~' && name[1]==0 && getenv("HOME"))
        strcpy(name, getenv("HOME"));
--- 117,123 ----
      if (*p) p++;
  
      if (name[0] == '.' && name[1] == 0)
!       getcwd(name, MAXPATHLEN);
  
      else if (name[0]=='~' && name[1]==0 && getenv("HOME"))
        strcpy(name, getenv("HOME"));
diff -c temp/scm/scmfig.h temp/nscm/scmfig.h
*** temp/scm/scmfig.h	Fri Sep 22 22:29:00 1995
--- temp/nscm/scmfig.h	Wed Mar 20 23:47:15 1996
***************
*** 50,55 ****
--- 50,59 ----
  #  include <strings.h>
  # endif
  
+ # ifndef HAVE_GETCWD
+ #  define getcwd(S,L) getwd(S)
+ # endif
+ 
  #else /* HAVE_CONFIG_H */
  
  # ifdef sequent
