diff -c orig/pngrutil.c ./pngrutil.c
*** orig/pngrutil.c	Mon May  1 12:43:38 1995
--- ./pngrutil.c	Mon Jun 26 22:29:18 1995
***************
*** 1362,1368 ****
  void png_read_finish_row(png_struct *png_ptr)
  {
     png_ptr->row_number++;
!    if (png_ptr->row_number < png_ptr->num_rows)
        return;
  
     if (png_ptr->interlaced)
--- 1362,1368 ----
  void png_read_finish_row(png_struct *png_ptr)
  {
     png_ptr->row_number++;
!    if (png_ptr->row_number < png_ptr->height)
        return;
  
     if (png_ptr->interlaced)
diff -c orig/pngstub.c ./pngstub.c
*** orig/pngstub.c	Mon May  1 15:58:42 1995
--- ./pngstub.c	Fri Jun 23 01:09:31 1995
***************
*** 195,201 ****
  void
  png_error(png_struct *png_ptr, char *message)
  {
!    fprintf(stderr, "pnglib error: %s\n", message);
  
     longjmp(png_ptr->jmpbuf, 1);
  }
--- 195,203 ----
  void
  png_error(png_struct *png_ptr, char *message)
  {
!    extern char zgv_pngerr[];
! 
!    sprintf(zgv_pngerr, "Error reading PNG file - %s", message);
  
     longjmp(png_ptr->jmpbuf, 1);
  }
***************
*** 210,215 ****
--- 212,219 ----
     if (!png_ptr)
        return;
  
+ #if 0
     fprintf(stderr, "pnglib warning: %s\n", message);
+ #endif
  }
  
