*** machdep.c.orig	Sat Oct  8 01:14:18 1994
--- machdep.c	Tue Dec  6 10:10:34 1994
***************
*** 111,116 ****
--- 111,118 ----
  #include "ether.h"
  #include "afsc.h"
  
+ #include "ppp.h"
+ 
  /* vm_map_t buffer_map; */
  extern vm_offset_t avail_end;
  
***************
*** 1250,1255 ****
--- 1252,1263 ----
  		netisr &= ~(1 << NETISR_ISO);
  		clnlintr();
  	}
+ #endif
+ #if NPPP > 0
+ 	if (netisr & (1 << NETISR_PPP)) {
+ 		netisr &= ~(1 << NETISR_PPP);
+ 		pppintr();
+ 	}
  #endif
  }
  
